{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6be1a832-ced2-4cfa-94bf-f23eb6558982",
   "metadata": {},
   "source": [
    "# <span style=\"color:green\">Formation South Green 2022</span> - Structural Variants Detection by using short and long reads \n",
    "\n",
    "# __DAY 2 : SNP calling__\n",
    "\n",
    "Created by C. Tranchant (DIADE-IRD), J. Orjuela (DIADE-IRD), F. Sabot (DIADE-IRD) and A. Dereeper (PHIM-IRD)\n",
    "\n",
    "## __1. Preparing the working environment__ \n",
    "\n",
    "### First create a dedicated folder to work "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fceac7fa-6c14-41f3-bccc-492edca10f06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAPPING-ILL  SV_DATA               training_SV_teaching_old\n",
      "MAPPING-ONT  training_SV_teaching\n"
     ]
    }
   ],
   "source": [
    "# go to work directory\n",
    "cd /home/jovyan/work/\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85dd67e1-6696-4c37-af1e-ca381a34fc68",
   "metadata": {},
   "source": [
    "### We are going to use only one clones to check is all is ok before run the whole of samples! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3f2f15f-a724-4e68-9eff-29d6a9995648",
   "metadata": {},
   "outputs": [],
   "source": [
    "## declare variables\n",
    "i=10\n",
    "REF_DIR=\"/home/jovyan/work/SV_DATA/REF/\"\n",
    "REF=\"/home/jovyan/work/SV_DATA/REF/reference.fasta\"\n",
    "ONT=\"/home/jovyan/work/SV_DATA/LONG_READS/Clone$i.fastq.gz\"\n",
    "ILL_R1=\"/home/jovyan/work/SV_DATA/SHORT_READS/Clone${i}_R1.fastq.gz\"\n",
    "ILL_R2=\"/home/jovyan/work/SV_DATA/SHORT_READS/Clone${i}_R2.fastq.gz\"\n",
    "BAM=\"/home/jovyan/work/MAPPING-ILL/dirClone10/Clone$i.SORTED.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1eb6344c-a5f2-4e0e-a798-6f0fbfab8483",
   "metadata": {},
   "outputs": [],
   "source": [
    "# go to SR mapping results\n",
    "cd /home/jovyan/work/MAPPING-ILL/dirClone${i}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9b436945-2f98-4662-85a2-da1dc1d904d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clone10.bam        Clone10.mappedpaired.flagstats  Clone10.SORTED.bam.bai\n",
      "Clone10.flagstats  Clone10.SORTED.bam\n"
     ]
    }
   ],
   "source": [
    "ls /home/jovyan/work/MAPPING-ILL/dirClone${i}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "597b1b89-f613-4094-a423-e486fe455491",
   "metadata": {},
   "source": [
    "### Mark duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "48bc398f-3783-4616-9c4a-faad14b818dc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MarkDuplicates in Clone10\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam -M duplicates.10.metrics -O Clone10.SORTED.MD.bam\n",
      "09:12:47.126 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:12:47 UTC 2022] MarkDuplicates --INPUT /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam --OUTPUT Clone10.SORTED.MD.bam --METRICS_FILE duplicates.10.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:12:47\tMarkDuplicates\tStart of doWork freeMemory: 76734352; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:12:47\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:12:47\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING: BAM index file /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam.bai is older than BAM /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam\n",
      "WARNING\t2022-06-17 09:12:47\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone1072030. Cause: String 'Reference-Clone1072030' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:12:49\tMarkDuplicates\tRead 218251 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:12:49\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 390348672; totalMemory: 689963008; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:12:49\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430738608; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tMarking 136 records as duplicates.\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "WARNING: BAM index file /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam.bai is older than BAM /home/jovyan/work/MAPPING-ILL/dirClone10/Clone10.SORTED.bam\n",
      "INFO\t2022-06-17 09:12:50\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tBefore output close freeMemory: 88355080; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:12:51\tMarkDuplicates\tAfter output close freeMemory: 84006920; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:12:51 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "echo -e \"\\nMarkDuplicates in Clone$i\";\n",
    "gatk MarkDuplicates -I $BAM -M duplicates.$i.metrics -O Clone$i.SORTED.MD.bam;\n",
    "samtools index Clone$i.SORTED.MD.bam;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6aafebd6-bc19-4494-8a57-197b68c1e8e0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 80100\n",
      "-rw-r--r-- 1 jovyan users 40019082 Jun 17 09:06 Clone10.bam\n",
      "-rw-r--r-- 1 jovyan users      381 Jun 17 09:06 Clone10.flagstats\n",
      "-rw-r--r-- 1 jovyan users      380 Jun 17 09:06 Clone10.mappedpaired.flagstats\n",
      "-rw-r--r-- 1 jovyan users 18215336 Jun 17 09:06 Clone10.SORTED.bam\n",
      "-rw-r--r-- 1 jovyan users     2856 Jun 16 14:29 Clone10.SORTED.bam.bai\n",
      "-rw-r--r-- 1 jovyan users 23758262 Jun 17 09:12 Clone10.SORTED.MD.bam\n",
      "-rw-r--r-- 1 jovyan users     2856 Jun 17 09:12 Clone10.SORTED.MD.bam.bai\n",
      "-rw-r--r-- 1 jovyan users     3466 Jun 17 09:12 duplicates.10.metrics\n"
     ]
    }
   ],
   "source": [
    "# check MD bam\n",
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8836759d-5737-4dba-ae5f-cf81a1485a8b",
   "metadata": {},
   "source": [
    "### Indexing reference with GATK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a944e78-53d4-44ea-93b8-9fc19b99219f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar CreateSequenceDictionary -R /home/jovyan/work/SV_DATA/REF/reference.fasta\n",
      "INFO\t2022-06-17 09:13:16\tCreateSequenceDictionary\tOutput dictionary will be written in /home/jovyan/work/SV_DATA/REF/reference.dict\n",
      "09:13:16.268 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:13:16 UTC 2022] CreateSequenceDictionary --REFERENCE /home/jovyan/work/SV_DATA/REF/reference.fasta --TRUNCATE_NAMES_AT_WHITESPACE true --NUM_SEQUENCES 2147483647 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "[Fri Jun 17 09:13:16 UTC 2022] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.00 minutes.\n",
      "Runtime.totalMemory()=633339904\n",
      "To get help, see http://broadinstitute.github.io/picard/index.html#GettingHelp\n",
      "picard.PicardException: /home/jovyan/work/SV_DATA/REF/reference.dict already exists.  Delete this file and try again, or specify a different output file.\n",
      "\tat picard.sam.CreateSequenceDictionary.doWork(CreateSequenceDictionary.java:220)\n",
      "\tat picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:308)\n",
      "\tat org.broadinstitute.hellbender.cmdline.PicardCommandLineProgramExecutor.instanceMain(PicardCommandLineProgramExecutor.java:37)\n",
      "\tat org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)\n",
      "\tat org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)\n",
      "\tat org.broadinstitute.hellbender.Main.main(Main.java:289)\n"
     ]
    },
    {
     "ename": "",
     "evalue": "3",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd $REF_DIR\n",
    "samtools faidx $REF\n",
    "gatk CreateSequenceDictionary -R $REF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc481e5-9697-4106-a145-bb7d6183a4a0",
   "metadata": {},
   "source": [
    "### SNP Calling for a Clone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5e8e1eb-3d05-40a5-9957-021f83ed9965",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calling Clone10\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4g -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar HaplotypeCaller --native-pair-hmm-threads 4 -I Clone10.SORTED.MD.bam -O Clone10.g.vcf -R /home/jovyan/work/SV_DATA/REF/reference.fasta -ERC GVCF\n",
      "09:13:42.117 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "09:13:42.452 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "09:13:42.453 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.6.1\n",
      "09:13:42.453 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "09:13:42.458 INFO  HaplotypeCaller - Initializing engine\n",
      "09:13:42.649 INFO  HaplotypeCaller - Done initializing engine\n",
      "09:13:42.651 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
      "09:13:42.656 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
      "09:13:42.657 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
      "09:13:42.657 INFO  HaplotypeCaller - Shutting down engine\n",
      "[June 17, 2022 at 9:13:42 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.01 minutes.\n",
      "Runtime.totalMemory()=632291328\n",
      "USAGE: HaplotypeCaller [arguments]\n",
      "\n",
      "Call germline SNPs and indels via local re-assembly of haplotypes\n",
      "Version:4.2.6.1\n",
      "\n",
      "\n",
      "Required Arguments:\n",
      "\n",
      "--input,-I <GATKPath>         BAM/SAM/CRAM file containing reads  This argument must be specified at least once.\n",
      "                              Required. \n",
      "\n",
      "--output,-O <GATKPath>        File to which variants should be written  Required. \n",
      "\n",
      "--reference,-R <GATKPath>     Reference sequence file  Required. \n",
      "\n",
      "\n",
      "Optional Arguments:\n",
      "\n",
      "--add-output-sam-program-record <Boolean>\n",
      "                              If true, adds a PG tag to created SAM/BAM/CRAM files.  Default value: true. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--add-output-vcf-command-line <Boolean>\n",
      "                              If true, adds a command line header line to created VCF files.  Default value: true.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--alleles <FeatureInput>      The set of alleles to force-call regardless of evidence  Default value: null. \n",
      "\n",
      "--annotate-with-num-discovered-alleles <Boolean>\n",
      "                              If provided, we will annotate records with the number of alternate alleles that were\n",
      "                              discovered (but not necessarily genotyped) at a given site  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--annotation,-A <String>      One or more specific annotations to add to variant calls  This argument may be specified 0\n",
      "                              or more times. Default value: null. Possible values: {AlleleFraction, AllelePseudoDepth,\n",
      "                              AS_BaseQualityRankSumTest, AS_FisherStrand, AS_InbreedingCoeff,\n",
      "                              AS_MappingQualityRankSumTest, AS_QualByDepth, AS_ReadPosRankSumTest, AS_RMSMappingQuality,\n",
      "                              AS_StrandBiasMutectAnnotation, AS_StrandOddsRatio, AssemblyComplexity, BaseQuality,\n",
      "                              BaseQualityHistogram, BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest,\n",
      "                              CountNs, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet,\n",
      "                              FeaturizedReadSets, FisherStrand, FragmentDepthPerAlleleBySample, FragmentLength,\n",
      "                              GenotypeSummaries, InbreedingCoeff, LikelihoodRankSumTest, MappingQuality,\n",
      "                              MappingQualityRankSumTest, MappingQualityZero, OrientationBiasReadCounts,\n",
      "                              OriginalAlignment, PossibleDeNovo, QualByDepth, ReadPosition, ReadPosRankSumTest,\n",
      "                              ReferenceBases, RMSMappingQuality, SampleList, StrandBiasBySample, StrandOddsRatio,\n",
      "                              TandemRepeat, UniqueAltReadCountPossible values: {\n",
      "\n",
      "--annotation-group,-G <String>One or more groups of annotations to apply to variant calls  This argument may be\n",
      "                              specified 0 or more times. Default value: null. Possible values:\n",
      "                              {AlleleSpecificAnnotation, AS_StandardAnnotation, GenotypeAnnotation, InfoFieldAnnotation,\n",
      "                              JumboGenotypeAnnotation, JumboInfoAnnotation, ReducibleAnnotation, StandardAnnotation,\n",
      "                              StandardHCAnnotation, StandardMutectAnnotation, VariantAnnotationPossible values: {\n",
      "\n",
      "--annotations-to-exclude,-AX <String>\n",
      "                              One or more specific annotations to exclude from variant calls  This argument may be\n",
      "                              specified 0 or more times. Default value: null. Possible values: {BaseQualityRankSumTest,\n",
      "                              ChromosomeCounts, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet,\n",
      "                              FisherStrand, InbreedingCoeff, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest,\n",
      "                              RMSMappingQuality, StrandOddsRatioPossible values: {\n",
      "\n",
      "--arguments_file <File>       read one or more arguments files and add them to the command line  This argument may be\n",
      "                              specified 0 or more times. Default value: null. \n",
      "\n",
      "--assembly-region-out <String>Output the assembly region to this IGV formatted file  Default value: null. \n",
      "\n",
      "--assembly-region-padding <Integer>\n",
      "                              Number of additional bases of context to include around each assembly region  Default\n",
      "                              value: 100. \n",
      "\n",
      "--base-quality-score-threshold <Byte>\n",
      "                              Base qualities below this threshold will be reduced to the minimum (6)  Default value: 18.\n",
      "\n",
      "--cloud-index-prefetch-buffer,-CIPB <Integer>\n",
      "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable). Defaults to\n",
      "                              cloudPrefetchBuffer if unset.  Default value: -1. \n",
      "\n",
      "--cloud-prefetch-buffer,-CPB <Integer>\n",
      "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable).  Default value: 40. \n",
      "\n",
      "--contamination-fraction-to-filter,-contamination <Double>\n",
      "                              Fraction of contamination in sequencing data (for all samples) to aggressively remove \n",
      "                              Default value: 0.0. \n",
      "\n",
      "--create-output-bam-index,-OBI <Boolean>\n",
      "                              If true, create a BAM/CRAM index when writing a coordinate-sorted BAM/CRAM file.  Default\n",
      "                              value: true. Possible values: {true, false} \n",
      "\n",
      "--create-output-bam-md5,-OBM <Boolean>\n",
      "                              If true, create a MD5 digest for any BAM/SAM/CRAM file created  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--create-output-variant-index,-OVI <Boolean>\n",
      "                              If true, create a VCF index when writing a coordinate-sorted VCF file.  Default value:\n",
      "                              true. Possible values: {true, false} \n",
      "\n",
      "--create-output-variant-md5,-OVM <Boolean>\n",
      "                              If true, create a a MD5 digest any VCF file created.  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--dbsnp,-D <FeatureInput>     dbSNP file  Default value: null. \n",
      "\n",
      "--disable-bam-index-caching,-DBIC <Boolean>\n",
      "                              If true, don't cache bam indexes, this will reduce memory requirements but may harm\n",
      "                              performance if many intervals are specified.  Caching is automatically disabled if there\n",
      "                              are no intervals specified.  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-read-filter,-DF <String>\n",
      "                              Read filters to be disabled before analysis  This argument may be specified 0 or more\n",
      "                              times. Default value: null. Possible values: {GoodCigarReadFilter, MappedReadFilter,\n",
      "                              MappingQualityAvailableReadFilter, MappingQualityReadFilter,\n",
      "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
      "                              NotSecondaryAlignmentReadFilter, PassesVendorQualityCheckReadFilter,\n",
      "                              WellformedReadFilterPossible values: {\n",
      "\n",
      "--disable-sequence-dictionary-validation <Boolean>\n",
      "                              If specified, do not check the sequence dictionaries from our inputs for compatibility.\n",
      "                              Use at your own risk!  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-use-dragstr-pair-hmm-scores <Boolean>\n",
      "                              disable DRAGstr pair-hmm score even when dragstr-params-path was provided  Default value:\n",
      "                              false. Possible values: {true, false} \n",
      "\n",
      "--dragen-mode <Boolean>       Single argument for enabling the bulk of DRAGEN-GATK features. NOTE: THIS WILL OVERWRITE\n",
      "                              PROVIDED ARGUMENT CHECK TOOL INFO TO SEE WHICH ARGUMENTS ARE SET).  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--dragstr-het-hom-ratio <Integer>\n",
      "                              het to hom prior ratio use with DRAGstr on  Default value: 2. \n",
      "\n",
      "--dragstr-params-path <GATKPath>\n",
      "                              location of the DRAGstr model parameters for STR error correction used in the Pair HMM.\n",
      "                              When provided, it overrides other PCR error correcting mechanisms  Default value: null. \n",
      "\n",
      "--enable-dynamic-read-disqualification-for-genotyping <Boolean>\n",
      "                              Will enable less strict read disqualification low base quality reads  Default value:\n",
      "                              false. Possible values: {true, false} \n",
      "\n",
      "--exclude-intervals,-XL <String>\n",
      "                              One or more genomic intervals to exclude from processing  This argument may be specified 0\n",
      "                              or more times. Default value: null. \n",
      "\n",
      "--founder-id <String>         Samples representing the population \"founders\"  This argument may be specified 0 or more\n",
      "                              times. Default value: null. \n",
      "\n",
      "--gatk-config-file <String>   A configuration file to use with the GATK.  Default value: null. \n",
      "\n",
      "--gcs-max-retries,-gcs-retries <Integer>\n",
      "                              If the GCS bucket channel errors out, how many times it will attempt to re-initiate the\n",
      "                              connection  Default value: 20. \n",
      "\n",
      "--gcs-project-for-requester-pays <String>\n",
      "                              Project to bill when accessing \"requester pays\" buckets. If unset, these buckets cannot be\n",
      "                              accessed.  User must have storage.buckets.get permission on the bucket being accessed. \n",
      "                              Default value: . \n",
      "\n",
      "--genotype-assignment-method,-gam <GenotypeAssignmentMethod>\n",
      "                              How we assign genotypes  Default value: USE_PLS_TO_ASSIGN. Possible values:\n",
      "                              {SET_TO_NO_CALL, USE_PLS_TO_ASSIGN, USE_POSTERIORS_ANNOTATION,\n",
      "                              SET_TO_NO_CALL_NO_ANNOTATIONS, BEST_MATCH_TO_ORIGINAL, DO_NOT_ASSIGN_GENOTYPES,\n",
      "                              USE_POSTERIOR_PROBABILITIES, PREFER_PLS} \n",
      "\n",
      "--graph-output,-graph <String>Write debug assembly graph information to this file  Default value: null. \n",
      "\n",
      "--help,-h <Boolean>           display the help message  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--heterozygosity <Double>     Heterozygosity value used to compute prior probabilities for any locus.  See the GATKDocs\n",
      "                              for full details on the meaning of this population genetics concept  Default value: 0.001.\n",
      "\n",
      "--heterozygosity-stdev <Double>\n",
      "                              Standard deviation of heterozygosity for SNP and indel calling.  Default value: 0.01. \n",
      "\n",
      "--indel-heterozygosity <Double>\n",
      "                              Heterozygosity for indel calling.  See the GATKDocs for heterozygosity for full details on\n",
      "                              the meaning of this population genetics concept  Default value: 1.25E-4. \n",
      "\n",
      "--interval-exclusion-padding,-ixp <Integer>\n",
      "                              Amount of padding (in bp) to add to each interval you are excluding.  Default value: 0. \n",
      "\n",
      "--interval-merging-rule,-imr <IntervalMergingRule>\n",
      "                              Interval merging rule for abutting intervals  Default value: ALL. Possible values: {ALL,\n",
      "                              OVERLAPPING_ONLY} \n",
      "\n",
      "--interval-padding,-ip <Integer>\n",
      "                              Amount of padding (in bp) to add to each interval you are including.  Default value: 0. \n",
      "\n",
      "--interval-set-rule,-isr <IntervalSetRule>\n",
      "                              Set merging approach to use for combining interval inputs  Default value: UNION. Possible\n",
      "                              values: {UNION, INTERSECTION} \n",
      "\n",
      "--intervals,-L <String>       One or more genomic intervals over which to operate  This argument may be specified 0 or\n",
      "                              more times. Default value: null. \n",
      "\n",
      "--lenient,-LE <Boolean>       Lenient processing of VCF files  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--max-assembly-region-size <Integer>\n",
      "                              Maximum size of an assembly region  Default value: 300. \n",
      "\n",
      "--max-reads-per-alignment-start <Integer>\n",
      "                              Maximum number of reads to retain per alignment start position. Reads above this threshold\n",
      "                              will be downsampled. Set to 0 to disable.  Default value: 50. \n",
      "\n",
      "--max-variants-per-shard <Integer>\n",
      "                              If non-zero, partitions VCF output into shards, each containing up to the given number of\n",
      "                              records.  Default value: 0. \n",
      "\n",
      "--min-assembly-region-size <Integer>\n",
      "                              Minimum size of an assembly region  Default value: 50. \n",
      "\n",
      "--min-base-quality-score,-mbq <Byte>\n",
      "                              Minimum base quality required to consider a base for calling  Default value: 10. \n",
      "\n",
      "--native-pair-hmm-threads <Integer>\n",
      "                              How many threads should a native pairHMM implementation use  Default value: 4. \n",
      "\n",
      "--native-pair-hmm-use-double-precision <Boolean>\n",
      "                              use double precision in the native pairHmm. This is slower but matches the java\n",
      "                              implementation better  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--num-reference-samples-if-no-call <Integer>\n",
      "                              Number of hom-ref genotypes to infer at sites not present in a panel  Default value: 0. \n",
      "\n",
      "--output-mode <OutputMode>    Specifies which type of calls we should output  Default value: EMIT_VARIANTS_ONLY.\n",
      "                              Possible values: {EMIT_VARIANTS_ONLY, EMIT_ALL_CONFIDENT_SITES, EMIT_ALL_ACTIVE_SITES} \n",
      "\n",
      "--pedigree,-ped <GATKPath>    Pedigree file for determining the population \"founders\"  Default value: null. \n",
      "\n",
      "--population-callset,-population <FeatureInput>\n",
      "                              Callset to use in calculating genotype priors  Default value: null. \n",
      "\n",
      "--QUIET <Boolean>             Whether to suppress job-summary info on System.err.  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--read-filter,-RF <String>    Read filters to be applied before analysis  This argument may be specified 0 or more\n",
      "                              times. Default value: null. Possible values: {AlignmentAgreesWithHeaderReadFilter,\n",
      "                              AllowAllReadsReadFilter, AmbiguousBaseReadFilter, CigarContainsNoNOperator,\n",
      "                              ExcessiveEndClippedReadFilter, FirstOfPairReadFilter, FragmentLengthReadFilter,\n",
      "                              GoodCigarReadFilter, HasReadGroupReadFilter, IntervalOverlapReadFilter, LibraryReadFilter,\n",
      "                              MappedReadFilter, MappingQualityAvailableReadFilter, MappingQualityNotZeroReadFilter,\n",
      "                              MappingQualityReadFilter, MatchingBasesAndQualsReadFilter, MateDifferentStrandReadFilter,\n",
      "                              MateDistantReadFilter, MateOnSameContigOrNoMappedMateReadFilter,\n",
      "                              MateUnmappedAndUnmappedReadFilter, MetricsReadFilter,\n",
      "                              NonChimericOriginalAlignmentReadFilter, NonZeroFragmentLengthReadFilter,\n",
      "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
      "                              NotOpticalDuplicateReadFilter, NotProperlyPairedReadFilter,\n",
      "                              NotSecondaryAlignmentReadFilter, NotSupplementaryAlignmentReadFilter,\n",
      "                              OverclippedReadFilter, PairedReadFilter, PassesVendorQualityCheckReadFilter,\n",
      "                              PlatformReadFilter, PlatformUnitReadFilter, PrimaryLineReadFilter,\n",
      "                              ProperlyPairedReadFilter, ReadGroupBlackListReadFilter, ReadGroupReadFilter,\n",
      "                              ReadLengthEqualsCigarLengthReadFilter, ReadLengthReadFilter, ReadNameReadFilter,\n",
      "                              ReadStrandFilter, SampleReadFilter, SecondOfPairReadFilter, SeqIsStoredReadFilter,\n",
      "                              SoftClippedReadFilter, ValidAlignmentEndReadFilter, ValidAlignmentStartReadFilter,\n",
      "                              WellformedReadFilterPossible values: {\n",
      "\n",
      "--read-index <GATKPath>       Indices to use for the read inputs. If specified, an index must be provided for every read\n",
      "                              input and in the same order as the read inputs. If this argument is not specified, the\n",
      "                              path to the index for each input will be inferred automatically.  This argument may be\n",
      "                              specified 0 or more times. Default value: null. \n",
      "\n",
      "--read-validation-stringency,-VS <ValidationStringency>\n",
      "                              Validation stringency for all SAM/BAM/CRAM/SRA files read by this program.  The default\n",
      "                              stringency value SILENT can improve performance when processing a BAM file in which\n",
      "                              variable-length data (read, qualities, tags) do not otherwise need to be decoded.  Default\n",
      "                              value: SILENT. Possible values: {STRICT, LENIENT, SILENT} \n",
      "\n",
      "--recover-dangling-heads <Boolean>\n",
      "                              This argument is deprecated since version 3.3  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--sample-name,-ALIAS <String> Name of single sample to use from a multi-sample bam  Default value: null. \n",
      "\n",
      "--sample-ploidy,-ploidy <Integer>\n",
      "                              Ploidy (number of chromosomes) per sample. For pooled data, set to (Number of samples in\n",
      "                              each pool * Sample Ploidy).  Default value: 2. \n",
      "\n",
      "--seconds-between-progress-updates <Double>\n",
      "                              Output traversal statistics every time this many seconds elapse  Default value: 10.0. \n",
      "\n",
      "--sequence-dictionary <GATKPath>\n",
      "                              Use the given sequence dictionary as the master/canonical sequence dictionary.  Must be a\n",
      "                              .dict file.  Default value: null. \n",
      "\n",
      "--sites-only-vcf-output <Boolean>\n",
      "                              If true, don't emit genotype fields when writing vcf file output.  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--standard-min-confidence-threshold-for-calling,-stand-call-conf <Double>\n",
      "                              The minimum phred-scaled confidence threshold at which variants should be called  Default\n",
      "                              value: 30.0. \n",
      "\n",
      "--tmp-dir <GATKPath>          Temp directory to use.  Default value: null. \n",
      "\n",
      "--use-jdk-deflater,-jdk-deflater <Boolean>\n",
      "                              Whether to use the JdkDeflater (as opposed to IntelDeflater)  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--use-jdk-inflater,-jdk-inflater <Boolean>\n",
      "                              Whether to use the JdkInflater (as opposed to IntelInflater)  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--use-new-qual-calculator,-new-qual <Boolean>\n",
      "                              Use the new AF model instead of the so-called exact model  Default value: true. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--use-posteriors-to-calculate-qual,-gp-qual <Boolean>\n",
      "                              if available, use the genotype posterior probabilities to calculate the site QUAL  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "--verbosity <LogLevel>        Control verbosity of logging.  Default value: INFO. Possible values: {ERROR, WARNING,\n",
      "                              INFO, DEBUG} \n",
      "\n",
      "--version <Boolean>           display the version number for this tool  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "\n",
      "Advanced Arguments:\n",
      "\n",
      "--active-probability-threshold <Double>\n",
      "                              Minimum probability for a locus to be considered active.  Default value: 0.002. \n",
      "\n",
      "--adaptive-pruning <Boolean>  Use Mutect2's adaptive graph pruning algorithm  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--adaptive-pruning-initial-error-rate <Double>\n",
      "                              Initial base error rate estimate for adaptive pruning  Default value: 0.001. \n",
      "\n",
      "--all-site-pls <Boolean>      Annotate all sites with PLs  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--allele-informative-reads-overlap-margin <Integer>\n",
      "                              Likelihood and read-based annotations will only take into consideration reads that overlap\n",
      "                              the variant or any base no further than this distance expressed in base pairs  Default\n",
      "                              value: 2. \n",
      "\n",
      "--allow-non-unique-kmers-in-ref <Boolean>\n",
      "                              Allow graphs that have non-unique kmers in the reference  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--apply-bqd <Boolean>         If enabled this argument will apply the DRAGEN-GATK BaseQualityDropout model to the\n",
      "                              genotyping model for filtering sites due to Linked Error mode.  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--apply-frd <Boolean>         If enabled this argument will apply the DRAGEN-GATK ForeignReadDetection model to the\n",
      "                              genotyping model for filtering sites.  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--bam-output,-bamout <String> File to which assembled haplotypes should be written  Default value: null. \n",
      "\n",
      "--bam-writer-type <WriterType>Which haplotypes should be written to the BAM  Default value: CALLED_HAPLOTYPES. Possible\n",
      "                              values: {ALL_POSSIBLE_HAPLOTYPES, CALLED_HAPLOTYPES, NO_HAPLOTYPES} \n",
      "\n",
      "--comparison,-comp <FeatureInput>\n",
      "                              Comparison VCF file(s)  This argument may be specified 0 or more times. Default value:\n",
      "                              null. \n",
      "\n",
      "--contamination-fraction-per-sample-file,-contamination-file <File>\n",
      "                              Tab-separated File containing fraction of contamination in sequencing data (per sample) to\n",
      "                              aggressively remove. Format should be \"<SampleID><TAB><Contamination>\" (Contamination is\n",
      "                              double) per line; No header.  Default value: null. \n",
      "\n",
      "--debug-assembly,-debug <Boolean>\n",
      "                              Print out verbose debug information about each assembly region  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--disable-cap-base-qualities-to-map-quality <Boolean>\n",
      "                              If false this disables capping of base qualities in the HMM to the mapping quality of the\n",
      "                              read  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-optimizations <Boolean>\n",
      "                              Don't skip calculations in ActiveRegions with no variants  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--disable-spanning-event-genotyping <Boolean>\n",
      "                              If enabled this argument will disable inclusion of the '*' spanning event when genotyping\n",
      "                              events that overlap deletions  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-symmetric-hmm-normalizing <Boolean>\n",
      "                              Toggle to revive legacy behavior of asymmetrically normalizing the arguments to the\n",
      "                              reference haplotype  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-tool-default-annotations <Boolean>\n",
      "                              Disable all tool default annotations  Default value: false. Possible values: {true, false}\n",
      "\n",
      "--disable-tool-default-read-filters <Boolean>\n",
      "                              Disable all tool default read filters (WARNING: many tools will not function correctly\n",
      "                              without their default read filters on)  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--do-not-correct-overlapping-quality <Boolean>\n",
      "                              Disable overlapping base quality correction  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--do-not-run-physical-phasing <Boolean>\n",
      "                              Disable physical phasing  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-increase-kmer-sizes-for-cycles <Boolean>\n",
      "                              Disable iterating over kmer sizes when graph cycles are detected  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--dont-use-dragstr-priors <Boolean>\n",
      "                              Forfeit the use of the DRAGstr model to calculate genotype priors. This argument does not\n",
      "                              have any effect in the absence of DRAGstr model parameters (--dragstr-model-params) \n",
      "                              Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-use-soft-clipped-bases <Boolean>\n",
      "                              Do not analyze soft clipped bases in the reads  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--emit-ref-confidence,-ERC <ReferenceConfidenceMode>\n",
      "                              Mode for emitting reference confidence scores (For Mutect2, this is a BETA feature) \n",
      "                              Default value: NONE. Possible values: {NONE, BP_RESOLUTION, GVCF} \n",
      "\n",
      "--enable-all-annotations <Boolean>\n",
      "                              Use all possible annotations (not for the faint of heart)  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--expected-mismatch-rate-for-read-disqualification <Double>\n",
      "                              Error rate used to set expectation for post HMM read disqualification based on mismatches \n",
      "                              Default value: 0.02. \n",
      "\n",
      "--floor-blocks <Boolean>      Output the band lower bound for each GQ block regardless of the data it represents \n",
      "                              Default value: false. Possible values: {true, false} \n",
      "\n",
      "--force-active <Boolean>      If provided, all regions will be marked as active  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--force-call-filtered-alleles,-genotype-filtered-alleles <Boolean>\n",
      "                              Force-call filtered alleles included in the resource specified by --alleles  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "--gvcf-gq-bands,-GQB <Integer>Exclusive upper bounds for reference confidence GQ bands (must be in [1, 100] and\n",
      "                              specified in increasing order)  This argument may be specified 0 or more times. Default\n",
      "                              value: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,\n",
      "                              24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,\n",
      "                              46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 70, 80, 90, 99]. \n",
      "\n",
      "--indel-size-to-eliminate-in-ref-model <Integer>\n",
      "                              The size of an indel to check for in the reference model  Default value: 10. \n",
      "\n",
      "--kmer-size <Integer>         Kmer size to use in the read threading assembler  This argument may be specified 0 or more\n",
      "                              times. Default value: [10, 25]. \n",
      "\n",
      "--linked-de-bruijn-graph <Boolean>\n",
      "                              If enabled, the Assembly Engine will construct a Linked De Bruijn graph to recover better\n",
      "                              haplotypes  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--mapping-quality-threshold-for-genotyping <Integer>\n",
      "                              Control the threshold for discounting reads from the genotyper due to mapping quality\n",
      "                              after the active region detection and assembly steps but before genotyping. NOTE: this is\n",
      "                              in contrast to the --minimum-mapping-quality argument which filters reads from all parts\n",
      "                              of the HaplotypeCaller. If you would like to call genotypes with a different threshold\n",
      "                              both arguments must be set.  Default value: 20. \n",
      "\n",
      "--max-alternate-alleles <Integer>\n",
      "                              Maximum number of alternate alleles to genotype  Default value: 6. \n",
      "\n",
      "--max-effective-depth-adjustment-for-frd <Integer>\n",
      "                              Set the maximum depth to modify FRD adjustment to in the event of high depth sites (0 to\n",
      "                              disable)  Default value: 0. \n",
      "\n",
      "--max-genotype-count <Integer>Maximum number of genotypes to consider at any site  Default value: 1024. \n",
      "\n",
      "--max-mnp-distance,-mnp-dist <Integer>\n",
      "                              Two or more phased substitutions separated by this distance or less are merged into MNPs. \n",
      "                              Default value: 0. \n",
      "\n",
      "--max-num-haplotypes-in-population <Integer>\n",
      "                              Maximum number of haplotypes to consider for your population  Default value: 128. \n",
      "\n",
      "--max-prob-propagation-distance <Integer>\n",
      "                              Upper limit on how many bases away probability mass can be moved around when calculating\n",
      "                              the boundaries between active and inactive assembly regions  Default value: 50. \n",
      "\n",
      "--max-unpruned-variants <Integer>\n",
      "                              Maximum number of variants in graph the adaptive pruner will allow  Default value: 100. \n",
      "\n",
      "--min-dangling-branch-length <Integer>\n",
      "                              Minimum length of a dangling branch to attempt recovery  Default value: 4. \n",
      "\n",
      "--min-pruning <Integer>       Minimum support to not prune paths in the graph  Default value: 2. \n",
      "\n",
      "--num-pruning-samples <Integer>\n",
      "                              Number of samples that must pass the minPruning threshold  Default value: 1. \n",
      "\n",
      "--pair-hmm-gap-continuation-penalty <Integer>\n",
      "                              Flat gap continuation penalty for use in the Pair HMM  Default value: 10. \n",
      "\n",
      "--pair-hmm-implementation,-pairHMM <Implementation>\n",
      "                              The PairHMM implementation to use for genotype likelihood calculations  Default value:\n",
      "                              FASTEST_AVAILABLE. Possible values: {EXACT, ORIGINAL, LOGLESS_CACHING,\n",
      "                              AVX_LOGLESS_CACHING, AVX_LOGLESS_CACHING_OMP, FASTEST_AVAILABLE} \n",
      "\n",
      "--pair-hmm-results-file <GATKPath>\n",
      "                              File to write exact pairHMM inputs/outputs to for debugging purposes  Default value: null.\n",
      "\n",
      "--pcr-indel-model <PCRErrorModel>\n",
      "                              The PCR indel model to use  Default value: CONSERVATIVE. Possible values: {NONE, HOSTILE,\n",
      "                              AGGRESSIVE, CONSERVATIVE} \n",
      "\n",
      "--phred-scaled-global-read-mismapping-rate <Integer>\n",
      "                              The global assumed mismapping rate for reads  Default value: 45. \n",
      "\n",
      "--pruning-lod-threshold <Double>\n",
      "                              Ln likelihood ratio threshold for adaptive pruning algorithm  Default value:\n",
      "                              2.302585092994046. \n",
      "\n",
      "--pruning-seeding-lod-threshold <Double>\n",
      "                              Ln likelihood ratio threshold for seeding subgraph of good variation in adaptive pruning\n",
      "                              algorithm  Default value: 9.210340371976184. \n",
      "\n",
      "--recover-all-dangling-branches <Boolean>\n",
      "                              Recover all dangling branches  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--showHidden <Boolean>        display hidden arguments  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--smith-waterman <Implementation>\n",
      "                              Which Smith-Waterman implementation to use, generally FASTEST_AVAILABLE is the right\n",
      "                              choice  Default value: JAVA. Possible values: {FASTEST_AVAILABLE, AVX_ENABLED, JAVA} \n",
      "\n",
      "--smith-waterman-dangling-end-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for dangling-end recovery.  Default value: -6. \n",
      "\n",
      "--smith-waterman-dangling-end-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for dangling-end recovery.  Default value: -110. \n",
      "\n",
      "--smith-waterman-dangling-end-match-value <Integer>\n",
      "                              Smith-Waterman match value for dangling-end recovery.  Default value: 25. \n",
      "\n",
      "--smith-waterman-dangling-end-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for dangling-end recovery.  Default value: -50. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -11. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -260. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-match-value <Integer>\n",
      "                              Smith-Waterman match value for haplotype-to-reference alignment.  Default value: 200. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -150. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for read-to-haplotype alignment.  Default value: -5. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for read-to-haplotype alignment.  Default value: -30. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-match-value <Integer>\n",
      "                              Smith-Waterman match value for read-to-haplotype alignment.  Default value: 10. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for read-to-haplotype alignment.  Default value: -15. \n",
      "\n",
      "--soft-clip-low-quality-ends <Boolean>\n",
      "                              If enabled will preserve low-quality read ends as softclips (used for DRAGEN-GATK BQD\n",
      "                              genotyper model)  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--transform-dragen-mapping-quality <Boolean>\n",
      "                              If enabled this argument will map DRAGEN aligner aligned reads with mapping quality <=250\n",
      "                              to scale up to MQ 50  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--use-filtered-reads-for-annotations <Boolean>\n",
      "                              Use the contamination-filtered read maps for the purposes of annotating variants  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "Conditional Arguments for annotation:\n",
      "\n",
      "Valid only if \"RMSMappingQuality\" is specified:\n",
      "--allow-old-rms-mapping-quality-annotation-data <Boolean>\n",
      "                              Override to allow old RMSMappingQuality annotated VCFs to function  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "Conditional Arguments for readFilter:\n",
      "\n",
      "Valid only if \"MappingQualityReadFilter\" is specified:\n",
      "--maximum-mapping-quality <Integer>\n",
      "                              Maximum mapping quality to keep (inclusive)  Default value: null. \n",
      "\n",
      "--minimum-mapping-quality <Integer>\n",
      "                              Minimum mapping quality to keep (inclusive)  Default value: 20. \n",
      "\n",
      "\n",
      "***********************************************************************\n",
      "\n",
      "A USER ERROR has occurred: Argument emit-ref-confidence has a bad value: Can only be used in single sample mode currently. Use the --sample-name argument to run on a single sample out of a multi-sample BAM file.\n",
      "\n",
      "***********************************************************************\n",
      "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# go to bam repertory\n",
    "cd /home/jovyan/work/MAPPING-ILL/dirClone${i}\n",
    "# change BAM file for MD one\n",
    "BAM=\"/home/jovyan/work/MAPPING-ILL/dirClone10/Clone$i.SORTED.MD.bam\"\n",
    "# lauch GATK HaplotypeCaller\n",
    "echo -e \"\\nCalling Clone$i\";\n",
    "gatk --java-options \"-Xmx4g\" HaplotypeCaller --native-pair-hmm-threads 4 -I Clone$i.SORTED.MD.bam -O Clone$i.g.vcf -R $REF -ERC GVCF;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d8ada194-7365-4886-9982-97c860842d64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "head: cannot open 'Clone10.g.vcf' for reading: No such file or directory\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "head Clone$i.g.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86d95ae1-5e20-4949-bc4f-a95fb3bb8cc7",
   "metadata": {},
   "source": [
    "### Error produced here is normal ! RG is important in mapping step  !! "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2851f135-79da-4868-aab4-6a8a1c956cc6",
   "metadata": {},
   "source": [
    "Yes, you have to relaunch mapping with all samples ! BUT ...\n",
    "\n",
    "Don't worry, bam files with correct `-R \"@RG\\tID:Clone${i}\\tSM:Clone${i}\"` parameter in bwa-mem2 command are available for downloading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "13993511-2f8a-4ebd-8b2c-dd0e82086632",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BAM_ILL/\n",
      "BAM_ILL/Clone10.SORTED.bam\n",
      "BAM_ILL/Clone11.SORTED.bam\n",
      "BAM_ILL/Clone12.SORTED.bam\n",
      "BAM_ILL/Clone13.SORTED.bam\n",
      "BAM_ILL/Clone14.SORTED.bam\n",
      "BAM_ILL/Clone15.SORTED.bam\n",
      "BAM_ILL/Clone16.SORTED.bam\n",
      "BAM_ILL/Clone17.SORTED.bam\n",
      "BAM_ILL/Clone18.SORTED.bam\n",
      "BAM_ILL/Clone19.SORTED.bam\n",
      "BAM_ILL/Clone1.SORTED.bam\n",
      "BAM_ILL/Clone20.SORTED.bam\n",
      "BAM_ILL/Clone2.SORTED.bam\n",
      "BAM_ILL/Clone3.SORTED.bam\n",
      "BAM_ILL/Clone4.SORTED.bam\n",
      "BAM_ILL/Clone5.SORTED.bam\n",
      "BAM_ILL/Clone6.SORTED.bam\n",
      "BAM_ILL/Clone7.SORTED.bam\n",
      "BAM_ILL/Clone8.SORTED.bam\n",
      "BAM_ILL/Clone9.SORTED.bam\n"
     ]
    }
   ],
   "source": [
    "cd /home/jovyan/work/\n",
    "wget --no-check-certificat -rm -nH --cut-dirs=1 --reject=\"index.html*\" wget https://itrop.ird.fr/sv-training/BAM_ILL.tar.gz\n",
    "tar zxvf BAM_ILL.tar.gz\n",
    "BAM_ILL=\"/home/jovyan/work/BAM_ILL\"\n",
    "rm BAM_ILL.tar.gz\n",
    "ls $BAM_ILL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d112c9-8b3d-444b-b516-46301b980de7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#wget --no-check-certificat -rm -nH --cut-dirs=1 --reject=\"index.html*\" wget https://itrop.ird.fr/sv-training/BAM_ONT.tar.gz\n",
    "#tar zxvf BAM_ONT.tar.gz\n",
    "#BAM_ONT=\"/home/jovyan/work/BAM_ONT\"\n",
    "#rm BAM_ONT.tar.gz\n",
    "#ls $BAM_ONT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74813193-5d58-4a21-ac39-02233366c7f4",
   "metadata": {},
   "source": [
    "## Calling all samples on one raw VCF with correct BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "39ce7d41-bdbc-40f9-8fbe-09eda8ce5f85",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clone10.SORTED.bam         Clone18.SORTED.MD.bam      Clone7.SORTED.bam\n",
      "Clone10.SORTED.bam.bai     Clone18.SORTED.MD.bam.bai  Clone7.SORTED.bam.bai\n",
      "Clone10.SORTED.MD.bam      Clone19.SORTED.bam         Clone7.SORTED.MD.bam\n",
      "Clone10.SORTED.MD.bam.bai  Clone19.SORTED.bam.bai     Clone7.SORTED.MD.bam.bai\n",
      "Clone11.SORTED.bam         Clone19.SORTED.MD.bam      Clone8.SORTED.bam\n",
      "Clone11.SORTED.bam.bai     Clone19.SORTED.MD.bam.bai  Clone8.SORTED.bam.bai\n",
      "Clone11.SORTED.MD.bam      Clone1.SORTED.bam          Clone8.SORTED.MD.bam\n",
      "Clone11.SORTED.MD.bam.bai  Clone1.SORTED.bam.bai      Clone8.SORTED.MD.bam.bai\n",
      "Clone12.SORTED.bam         Clone1.SORTED.MD.bam       Clone9.SORTED.bam\n",
      "Clone12.SORTED.bam.bai     Clone1.SORTED.MD.bam.bai   Clone9.SORTED.bam.bai\n",
      "Clone12.SORTED.MD.bam      Clone20.SORTED.bam         Clone9.SORTED.MD.bam\n",
      "Clone12.SORTED.MD.bam.bai  Clone20.SORTED.bam.bai     Clone9.SORTED.MD.bam.bai\n",
      "Clone13.SORTED.bam         Clone20.SORTED.MD.bam      duplicates.10.metrics\n",
      "Clone13.SORTED.bam.bai     Clone20.SORTED.MD.bam.bai  duplicates.11.metrics\n",
      "Clone13.SORTED.MD.bam      Clone2.SORTED.bam          duplicates.12.metrics\n",
      "Clone13.SORTED.MD.bam.bai  Clone2.SORTED.bam.bai      duplicates.13.metrics\n",
      "Clone14.SORTED.bam         Clone2.SORTED.MD.bam       duplicates.14.metrics\n",
      "Clone14.SORTED.bam.bai     Clone2.SORTED.MD.bam.bai   duplicates.15.metrics\n",
      "Clone14.SORTED.MD.bam      Clone3.SORTED.bam          duplicates.16.metrics\n",
      "Clone14.SORTED.MD.bam.bai  Clone3.SORTED.bam.bai      duplicates.17.metrics\n",
      "Clone15.SORTED.bam         Clone3.SORTED.MD.bam       duplicates.18.metrics\n",
      "Clone15.SORTED.bam.bai     Clone3.SORTED.MD.bam.bai   duplicates.19.metrics\n",
      "Clone15.SORTED.MD.bam      Clone4.SORTED.bam          duplicates.1.metrics\n",
      "Clone15.SORTED.MD.bam.bai  Clone4.SORTED.bam.bai      duplicates.20.metrics\n",
      "Clone16.SORTED.bam         Clone4.SORTED.MD.bam       duplicates.2.metrics\n",
      "Clone16.SORTED.bam.bai     Clone4.SORTED.MD.bam.bai   duplicates.3.metrics\n",
      "Clone16.SORTED.MD.bam      Clone5.SORTED.bam          duplicates.4.metrics\n",
      "Clone16.SORTED.MD.bam.bai  Clone5.SORTED.bam.bai      duplicates.5.metrics\n",
      "Clone17.SORTED.bam         Clone5.SORTED.MD.bam       duplicates.6.metrics\n",
      "Clone17.SORTED.bam.bai     Clone5.SORTED.MD.bam.bai   duplicates.7.metrics\n",
      "Clone17.SORTED.MD.bam      Clone6.SORTED.bam          duplicates.8.metrics\n",
      "Clone17.SORTED.MD.bam.bai  Clone6.SORTED.bam.bai      duplicates.9.metrics\n",
      "Clone18.SORTED.bam         Clone6.SORTED.MD.bam\n",
      "Clone18.SORTED.bam.bai     Clone6.SORTED.MD.bam.bai\n"
     ]
    }
   ],
   "source": [
    "BAM_ILL=\"/home/jovyan/work/BAM_ILL\"\n",
    "ls $BAM_ILL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fc1f5a57-96ff-412c-a866-f536314c1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a work directory to SNP calling results\n",
    "mkdir -p /home/jovyan/work/VCF\n",
    "VCF_DIR=\"/home/jovyan/work/VCF\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a1a59ea-cd90-4533-a33d-17e1db77a17a",
   "metadata": {},
   "source": [
    "### Markduplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f16f3738-431c-44c7-a4f6-e4493f92a232",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MarkDuplicates in Clone1\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone1.SORTED.bam -M duplicates.1.metrics -O Clone1.SORTED.MD.bam\n",
      "09:33:18.208 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:33:18 UTC 2022] MarkDuplicates --INPUT Clone1.SORTED.bam --OUTPUT Clone1.SORTED.MD.bam --METRICS_FILE duplicates.1.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:33:18\tMarkDuplicates\tStart of doWork freeMemory: 70486640; totalMemory: 98566144; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:18\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:33:18\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:33:18\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone1198966. Cause: String 'Reference-Clone1198966' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:33:20\tMarkDuplicates\tRead 266640 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:33:21\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 422854688; totalMemory: 727711744; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:21\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:33:21\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:21\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:22\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:33:22\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1428641232; totalMemory: 3563061248; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:22\tMarkDuplicates\tMarking 184 records as duplicates.\n",
      "INFO\t2022-06-17 09:33:22\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:33:22\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:33:23\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:33:23\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:33:23\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:33:23\tMarkDuplicates\tBefore output close freeMemory: 94628040; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:23\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:33:24\tMarkDuplicates\tAfter output close freeMemory: 75617384; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:33:24 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.10 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone2\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone2.SORTED.bam -M duplicates.2.metrics -O Clone2.SORTED.MD.bam\n",
      "09:33:28.006 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:33:28 UTC 2022] MarkDuplicates --INPUT Clone2.SORTED.bam --OUTPUT Clone2.SORTED.MD.bam --METRICS_FILE duplicates.2.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:33:28\tMarkDuplicates\tStart of doWork freeMemory: 85082016; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:28\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:33:28\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:33:28\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone2111268. Cause: String 'Reference-Clone2111268' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:33:30\tMarkDuplicates\tRead 266640 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:33:30\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 401882800; totalMemory: 706740224; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:30\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737944; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tMarking 176 records as duplicates.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:33:31\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tBefore output close freeMemory: 851921944; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:33:33\tMarkDuplicates\tAfter output close freeMemory: 84005544; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:33:33 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone3\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone3.SORTED.bam -M duplicates.3.metrics -O Clone3.SORTED.MD.bam\n",
      "09:33:36.932 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:33:36 UTC 2022] MarkDuplicates --INPUT Clone3.SORTED.bam --OUTPUT Clone3.SORTED.MD.bam --METRICS_FILE duplicates.3.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:33:37\tMarkDuplicates\tStart of doWork freeMemory: 76736904; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:37\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:33:37\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:33:37\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone394898. Cause: String 'Reference-Clone394898' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:33:39\tMarkDuplicates\tRead 266638 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:33:39\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 336870248; totalMemory: 641728512; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:39\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:33:39\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434931192; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tMarking 174 records as duplicates.\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:33:40\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tBefore output close freeMemory: 103023616; totalMemory: 134217728; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:33:41\tMarkDuplicates\tAfter output close freeMemory: 75653856; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:33:41 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone4\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone4.SORTED.bam -M duplicates.4.metrics -O Clone4.SORTED.MD.bam\n",
      "09:33:45.333 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:33:45 UTC 2022] MarkDuplicates --INPUT Clone4.SORTED.bam --OUTPUT Clone4.SORTED.MD.bam --METRICS_FILE duplicates.4.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:33:45\tMarkDuplicates\tStart of doWork freeMemory: 76740448; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:45\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:33:45\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:33:45\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone4256640. Cause: String 'Reference-Clone4256640' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:33:47\tMarkDuplicates\tRead 265394 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:33:47\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 253119672; totalMemory: 557842432; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:47\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737392; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tMarking 174 records as duplicates.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:33:48\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tBefore output close freeMemory: 94360208; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:33:50\tMarkDuplicates\tAfter output close freeMemory: 69399480; totalMemory: 98566144; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:33:50 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=98566144\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone5\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone5.SORTED.bam -M duplicates.5.metrics -O Clone5.SORTED.MD.bam\n",
      "09:33:54.529 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:33:54 UTC 2022] MarkDuplicates --INPUT Clone5.SORTED.bam --OUTPUT Clone5.SORTED.MD.bam --METRICS_FILE duplicates.5.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:33:54\tMarkDuplicates\tStart of doWork freeMemory: 76736160; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:54\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:33:54\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:33:54\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone5186012. Cause: String 'Reference-Clone5186012' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:33:56\tMarkDuplicates\tRead 266140 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:33:56\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 500502024; totalMemory: 805306368; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:56\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737096; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tMarking 184 records as duplicates.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:33:57\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tBefore output close freeMemory: 83934312; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:33:59\tMarkDuplicates\tAfter output close freeMemory: 84005008; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:33:59 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone6\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone6.SORTED.bam -M duplicates.6.metrics -O Clone6.SORTED.MD.bam\n",
      "09:34:03.444 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:03 UTC 2022] MarkDuplicates --INPUT Clone6.SORTED.bam --OUTPUT Clone6.SORTED.MD.bam --METRICS_FILE duplicates.6.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:03\tMarkDuplicates\tStart of doWork freeMemory: 76734968; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:03\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:03\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:03\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone6308622. Cause: String 'Reference-Clone6308622' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:06\tMarkDuplicates\tRead 265622 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:06\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 246803024; totalMemory: 551550976; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:06\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434931776; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tMarking 174 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:07\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:08\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:08\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:08\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:09\tMarkDuplicates\tBefore output close freeMemory: 88224512; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:09\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:09\tMarkDuplicates\tAfter output close freeMemory: 69401216; totalMemory: 98566144; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:09 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=98566144\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone7\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone7.SORTED.bam -M duplicates.7.metrics -O Clone7.SORTED.MD.bam\n",
      "09:34:12.975 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:13 UTC 2022] MarkDuplicates --INPUT Clone7.SORTED.bam --OUTPUT Clone7.SORTED.MD.bam --METRICS_FILE duplicates.7.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:13\tMarkDuplicates\tStart of doWork freeMemory: 85081408; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:13\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:13\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:13\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone7215560. Cause: String 'Reference-Clone7215560' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:15\tMarkDuplicates\tRead 264814 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:15\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 280443784; totalMemory: 585105408; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:15\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430736608; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tMarking 198 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:16\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tBefore output close freeMemory: 94620784; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:18\tMarkDuplicates\tAfter output close freeMemory: 69402224; totalMemory: 98566144; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:18 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=98566144\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone8\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone8.SORTED.bam -M duplicates.8.metrics -O Clone8.SORTED.MD.bam\n",
      "09:34:22.621 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:22 UTC 2022] MarkDuplicates --INPUT Clone8.SORTED.bam --OUTPUT Clone8.SORTED.MD.bam --METRICS_FILE duplicates.8.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:22\tMarkDuplicates\tStart of doWork freeMemory: 76734744; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:22\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:22\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:22\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone8347356. Cause: String 'Reference-Clone8347356' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:25\tMarkDuplicates\tRead 262530 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:25\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 358286328; totalMemory: 662700032; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:25\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434930592; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tMarking 188 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:26\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tBefore output close freeMemory: 89463472; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:28\tMarkDuplicates\tAfter output close freeMemory: 75615280; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:28 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone9\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone9.SORTED.bam -M duplicates.9.metrics -O Clone9.SORTED.MD.bam\n",
      "09:34:32.020 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:32 UTC 2022] MarkDuplicates --INPUT Clone9.SORTED.bam --OUTPUT Clone9.SORTED.MD.bam --METRICS_FILE duplicates.9.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:32\tMarkDuplicates\tStart of doWork freeMemory: 76735432; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:32\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:32\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:32\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone9279986. Cause: String 'Reference-Clone9279986' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:34\tMarkDuplicates\tRead 250852 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:34\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 338583040; totalMemory: 641728512; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:34\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434931536; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tMarking 180 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:35\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tBefore output close freeMemory: 102953752; totalMemory: 134217728; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:36\tMarkDuplicates\tAfter output close freeMemory: 75653448; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:36 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone10\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone10.SORTED.bam -M duplicates.10.metrics -O Clone10.SORTED.MD.bam\n",
      "09:34:40.444 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:40 UTC 2022] MarkDuplicates --INPUT Clone10.SORTED.bam --OUTPUT Clone10.SORTED.MD.bam --METRICS_FILE duplicates.10.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:40\tMarkDuplicates\tStart of doWork freeMemory: 76735344; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:40\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:40\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:40\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone1072030. Cause: String 'Reference-Clone1072030' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:43\tMarkDuplicates\tRead 218251 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:43\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 289684416; totalMemory: 589299712; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:43\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:43\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434931952; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tMarking 136 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:44\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tBefore output close freeMemory: 88489200; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:45\tMarkDuplicates\tAfter output close freeMemory: 84005344; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:45 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone11\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone11.SORTED.bam -M duplicates.11.metrics -O Clone11.SORTED.MD.bam\n",
      "09:34:49.519 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:49 UTC 2022] MarkDuplicates --INPUT Clone11.SORTED.bam --OUTPUT Clone11.SORTED.MD.bam --METRICS_FILE duplicates.11.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:49\tMarkDuplicates\tStart of doWork freeMemory: 76735712; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:49\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:49\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:49\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone11156446. Cause: String 'Reference-Clone11156446' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:34:52\tMarkDuplicates\tRead 216546 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:34:52\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 352792704; totalMemory: 652214272; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:52\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737688; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tMarking 146 records as duplicates.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:34:53\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tBefore output close freeMemory: 90226128; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:34:54\tMarkDuplicates\tAfter output close freeMemory: 75655152; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:34:54 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone12\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone12.SORTED.bam -M duplicates.12.metrics -O Clone12.SORTED.MD.bam\n",
      "09:34:58.610 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:34:58 UTC 2022] MarkDuplicates --INPUT Clone12.SORTED.bam --OUTPUT Clone12.SORTED.MD.bam --METRICS_FILE duplicates.12.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:34:58\tMarkDuplicates\tStart of doWork freeMemory: 76737752; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:34:58\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:34:58\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:34:58\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone12194702. Cause: String 'Reference-Clone12194702' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:00\tMarkDuplicates\tRead 186906 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:00\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 290986624; totalMemory: 587202560; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:00\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:01\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:01\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:02\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:02\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1428640120; totalMemory: 3563061248; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:02\tMarkDuplicates\tMarking 126 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:02\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:02\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tBefore output close freeMemory: 1009207672; totalMemory: 3563061248; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:03\tMarkDuplicates\tAfter output close freeMemory: 90296736; totalMemory: 119537664; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:03 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=119537664\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone13\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone13.SORTED.bam -M duplicates.13.metrics -O Clone13.SORTED.MD.bam\n",
      "09:35:06.962 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:06 UTC 2022] MarkDuplicates --INPUT Clone13.SORTED.bam --OUTPUT Clone13.SORTED.MD.bam --METRICS_FILE duplicates.13.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:07\tMarkDuplicates\tStart of doWork freeMemory: 85083824; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:07\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:07\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:07\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone13233262. Cause: String 'Reference-Clone13233262' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:09\tMarkDuplicates\tRead 223351 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:09\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 385642248; totalMemory: 685768704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:09\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434930792; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tMarking 174 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:10\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tBefore output close freeMemory: 83933600; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:12\tMarkDuplicates\tAfter output close freeMemory: 75653496; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:12 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone14\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone14.SORTED.bam -M duplicates.14.metrics -O Clone14.SORTED.MD.bam\n",
      "09:35:16.035 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:16 UTC 2022] MarkDuplicates --INPUT Clone14.SORTED.bam --OUTPUT Clone14.SORTED.MD.bam --METRICS_FILE duplicates.14.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:16\tMarkDuplicates\tStart of doWork freeMemory: 76735096; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:16\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:16\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:16\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone14131008. Cause: String 'Reference-Clone14131008' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:18\tMarkDuplicates\tRead 222631 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:19\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 253638312; totalMemory: 553648128; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:19\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:19\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:19\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:19\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:20\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737864; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:20\tMarkDuplicates\tMarking 192 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:20\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:20\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tBefore output close freeMemory: 94653968; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:21\tMarkDuplicates\tAfter output close freeMemory: 75656912; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:21 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone15\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone15.SORTED.bam -M duplicates.15.metrics -O Clone15.SORTED.MD.bam\n",
      "09:35:24.880 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:24 UTC 2022] MarkDuplicates --INPUT Clone15.SORTED.bam --OUTPUT Clone15.SORTED.MD.bam --METRICS_FILE duplicates.15.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:25\tMarkDuplicates\tStart of doWork freeMemory: 76734592; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:25\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:25\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:25\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone15293750. Cause: String 'Reference-Clone15293750' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:27\tMarkDuplicates\tRead 218801 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:27\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 214302304; totalMemory: 513802240; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:27\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:27\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:27\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:28\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:28\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737120; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:28\tMarkDuplicates\tMarking 182 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:28\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:28\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tBefore output close freeMemory: 88216512; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:29\tMarkDuplicates\tAfter output close freeMemory: 69400648; totalMemory: 98566144; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:29 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=98566144\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone16\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone16.SORTED.bam -M duplicates.16.metrics -O Clone16.SORTED.MD.bam\n",
      "09:35:33.436 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:33 UTC 2022] MarkDuplicates --INPUT Clone16.SORTED.bam --OUTPUT Clone16.SORTED.MD.bam --METRICS_FILE duplicates.16.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:33\tMarkDuplicates\tStart of doWork freeMemory: 85081640; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:33\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:33\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:34\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone16157448. Cause: String 'Reference-Clone16157448' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:36\tMarkDuplicates\tRead 231835 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:36\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 278058712; totalMemory: 578813952; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:36\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1428639624; totalMemory: 3563061248; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tMarking 206 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:37\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tBefore output close freeMemory: 94561192; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:38\tMarkDuplicates\tAfter output close freeMemory: 84004768; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:38 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone17\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone17.SORTED.bam -M duplicates.17.metrics -O Clone17.SORTED.MD.bam\n",
      "09:35:42.458 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:42 UTC 2022] MarkDuplicates --INPUT Clone17.SORTED.bam --OUTPUT Clone17.SORTED.MD.bam --METRICS_FILE duplicates.17.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:42\tMarkDuplicates\tStart of doWork freeMemory: 76702424; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:42\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:42\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:42\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone17287778. Cause: String 'Reference-Clone17287778' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:45\tMarkDuplicates\tRead 220711 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:45\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 274827976; totalMemory: 574619648; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:45\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1439124648; totalMemory: 3573547008; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tMarking 182 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:46\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tBefore output close freeMemory: 83933104; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:47\tMarkDuplicates\tAfter output close freeMemory: 75651112; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:47 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone18\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone18.SORTED.bam -M duplicates.18.metrics -O Clone18.SORTED.MD.bam\n",
      "09:35:51.738 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:35:51 UTC 2022] MarkDuplicates --INPUT Clone18.SORTED.bam --OUTPUT Clone18.SORTED.MD.bam --METRICS_FILE duplicates.18.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:35:52\tMarkDuplicates\tStart of doWork freeMemory: 85081880; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:52\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:35:52\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:35:52\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone18279864. Cause: String 'Reference-Clone18279864' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:35:54\tMarkDuplicates\tRead 206831 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:35:54\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 356011960; totalMemory: 654311424; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:54\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430737848; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tMarking 168 records as duplicates.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:35:55\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tBefore output close freeMemory: 83585768; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:35:56\tMarkDuplicates\tAfter output close freeMemory: 84005544; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:35:56 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone19\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone19.SORTED.bam -M duplicates.19.metrics -O Clone19.SORTED.MD.bam\n",
      "09:36:00.338 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:36:00 UTC 2022] MarkDuplicates --INPUT Clone19.SORTED.bam --OUTPUT Clone19.SORTED.MD.bam --METRICS_FILE duplicates.19.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:36:00\tMarkDuplicates\tStart of doWork freeMemory: 85082704; totalMemory: 113246208; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:00\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:36:00\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:36:00\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone1940994. Cause: String 'Reference-Clone1940994' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:36:02\tMarkDuplicates\tRead 222447 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:36:03\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 410951136; totalMemory: 710934528; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:03\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:36:03\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1430738368; totalMemory: 3565158400; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tMarking 188 records as duplicates.\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:36:04\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tBefore output close freeMemory: 88213720; totalMemory: 119537664; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:36:05\tMarkDuplicates\tAfter output close freeMemory: 75652552; totalMemory: 104857600; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:36:05 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=104857600\n",
      "Tool returned:\n",
      "0\n",
      "\n",
      "MarkDuplicates in Clone20\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar MarkDuplicates -I Clone20.SORTED.bam -M duplicates.20.metrics -O Clone20.SORTED.MD.bam\n",
      "09:36:09.526 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jun 17 09:36:09 UTC 2022] MarkDuplicates --INPUT Clone20.SORTED.bam --OUTPUT Clone20.SORTED.MD.bam --METRICS_FILE duplicates.20.metrics --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --ADD_PG_TAG_TO_READS true --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "INFO\t2022-06-17 09:36:09\tMarkDuplicates\tStart of doWork freeMemory: 76697152; totalMemory: 104857600; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:09\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2022-06-17 09:36:09\tMarkDuplicates\tWill retain up to 30507482 data points before spilling to disk.\n",
      "WARNING\t2022-06-17 09:36:09\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: Reference-Clone20209626. Cause: String 'Reference-Clone20209626' did not start with a parsable number.\n",
      "INFO\t2022-06-17 09:36:12\tMarkDuplicates\tRead 213813 records. 0 pairs never matched.\n",
      "INFO\t2022-06-17 09:36:12\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 300734576; totalMemory: 599785472; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:12\tMarkDuplicates\tWill retain up to 263127040 duplicate indices before spilling to disk.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1434930968; totalMemory: 3569352704; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tMarking 146 records as duplicates.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2022-06-17 09:36:13\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tBefore output close freeMemory: 94966664; totalMemory: 125829120; maxMemory: 8420065280\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2022-06-17 09:36:14\tMarkDuplicates\tAfter output close freeMemory: 84004424; totalMemory: 113246208; maxMemory: 8420065280\n",
      "[Fri Jun 17 09:36:14 UTC 2022] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=113246208\n",
      "Tool returned:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "cd $BAM_ILL\n",
    "for i in {1..20}\n",
    "    do\n",
    "        samtools index Clone$i.SORTED.bam;\n",
    "        echo -e \"\\nMarkDuplicates in Clone$i\";\n",
    "        gatk MarkDuplicates -I Clone$i.SORTED.bam -M duplicates.$i.metrics -O Clone$i.SORTED.MD.bam;\n",
    "        samtools index Clone$i.SORTED.MD.bam;\n",
    "    done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88f51c68-5647-4675-81d0-75aa250eb6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $BAM_ILL\n",
    "for i in {1..20}\n",
    "    do\n",
    "        echo -e \"\\nMarkDuplicates in Clone$i\";\n",
    "        gatk MarkDuplicates -I Clone$i.SORTED.bam -M duplicates.$i.metrics -O Clone$i.SORTED.MD.bam;\n",
    "        samtools index Clone$i.SORTED.MD.bam;\n",
    "    done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0251d02d-1c88-4b9d-9f06-52f5cd2d841b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calling Clone1\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4g -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar HaplotypeCaller --native-pair-hmm-threads 4 -I Clone1.SORTED.MD.bam -O /home/jovyan/work/VCF/Clone1.g.vcf -R /home/jovyan/work/SV_DATA/REF/reference.fasta -ERC GVCF\n",
      "11:31:40.773 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "11:31:40.925 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "11:31:40.925 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.6.1\n",
      "11:31:40.925 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "11:31:40.930 INFO  HaplotypeCaller - Initializing engine\n",
      "11:31:41.101 INFO  HaplotypeCaller - Done initializing engine\n",
      "11:31:41.103 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
      "11:31:41.108 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
      "11:31:41.108 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
      "11:31:41.123 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
      "11:31:41.125 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
      "11:31:41.142 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
      "11:31:41.147 INFO  IntelPairHmm - Available threads: 8\n",
      "11:31:41.147 INFO  IntelPairHmm - Requested threads: 4\n",
      "11:31:41.148 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
      "11:31:41.210 INFO  ProgressMeter - Starting traversal\n",
      "11:31:41.210 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "11:31:41.656 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position Reference:39 and possibly subsequent; at least 10 samples must have called genotypes\n",
      "11:31:51.296 INFO  ProgressMeter -      Reference:38941              0.2                   350           2082.1\n",
      "11:32:01.880 INFO  ProgressMeter -      Reference:88316              0.3                   790           2293.2\n",
      "11:32:12.265 INFO  ProgressMeter -     Reference:138317              0.5                  1240           2395.7\n",
      "11:32:22.349 INFO  ProgressMeter -     Reference:188947              0.7                  1710           2494.0\n",
      "11:32:32.624 INFO  ProgressMeter -     Reference:246817              0.9                  2240           2614.1\n",
      "11:32:43.121 INFO  ProgressMeter -     Reference:287747              1.0                  2600           2519.7\n",
      "11:32:53.152 INFO  ProgressMeter -     Reference:411295              1.2                  3290           2743.9\n",
      "11:33:03.152 INFO  ProgressMeter -     Reference:457965              1.4                  3720           2723.9\n",
      "11:33:13.294 INFO  ProgressMeter -     Reference:512107              1.5                  4200           2736.6\n",
      "11:33:23.297 INFO  ProgressMeter -     Reference:557218              1.7                  4620           2715.3\n",
      "11:33:33.468 INFO  ProgressMeter -     Reference:610996              1.9                  5110           2731.2\n",
      "11:33:43.473 INFO  ProgressMeter -     Reference:652830              2.0                  5500           2699.1\n",
      "11:33:53.625 INFO  ProgressMeter -     Reference:706661              2.2                  5970           2705.1\n",
      "11:34:04.089 INFO  ProgressMeter -     Reference:750419              2.4                  6360           2670.8\n",
      "11:34:14.101 INFO  ProgressMeter -     Reference:810126              2.5                  6930           2719.6\n",
      "11:34:24.168 INFO  ProgressMeter -     Reference:854399              2.7                  7320           2695.2\n",
      "11:34:34.236 INFO  ProgressMeter -     Reference:903502              2.9                  7750           2687.5\n",
      "11:34:44.301 INFO  ProgressMeter -     Reference:971447              3.1                  8360           2739.6\n",
      "11:34:50.363 WARN  IntelInflater - Zero Bytes Written : 0\n",
      "11:34:50.457 INFO  HaplotypeCaller - 23964 read(s) filtered by: MappingQualityReadFilter \n",
      "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
      "0 read(s) filtered by: MappedReadFilter \n",
      "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
      "162 read(s) filtered by: NotDuplicateReadFilter \n",
      "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
      "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
      "0 read(s) filtered by: GoodCigarReadFilter \n",
      "0 read(s) filtered by: WellformedReadFilter \n",
      "24126 total reads filtered\n",
      "11:34:50.457 INFO  ProgressMeter -    Reference:1018030              3.2                  8686           2753.9\n",
      "11:34:50.457 INFO  ProgressMeter - Traversal complete. Processed 8686 total regions in 3.2 minutes.\n",
      "11:34:50.470 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.175308914\n",
      "11:34:50.470 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 14.157904148\n",
      "11:34:50.475 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 104.78 sec\n",
      "11:34:50.475 INFO  HaplotypeCaller - Shutting down engine\n",
      "[June 17, 2022 at 11:34:50 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 3.16 minutes.\n",
      "Runtime.totalMemory()=759169024\n",
      "\n",
      "Calling Clone2\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4g -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar HaplotypeCaller --native-pair-hmm-threads 4 -I Clone2.SORTED.MD.bam -O /home/jovyan/work/VCF/Clone2.g.vcf -R /home/jovyan/work/SV_DATA/REF/reference.fasta -ERC GVCF\n",
      "11:34:53.380 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "11:34:53.565 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "11:34:53.565 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.6.1\n",
      "11:34:53.566 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "11:34:53.570 INFO  HaplotypeCaller - Initializing engine\n",
      "11:34:53.792 INFO  HaplotypeCaller - Done initializing engine\n",
      "11:34:53.795 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
      "11:34:53.801 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
      "11:34:53.801 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
      "11:34:53.821 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
      "11:34:53.827 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
      "11:34:53.848 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
      "11:34:53.863 INFO  IntelPairHmm - Available threads: 8\n",
      "11:34:53.863 INFO  IntelPairHmm - Requested threads: 4\n",
      "11:34:53.863 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
      "11:34:53.938 INFO  ProgressMeter - Starting traversal\n",
      "11:34:53.938 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "11:34:54.520 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position Reference:69 and possibly subsequent; at least 10 samples must have called genotypes\n",
      "11:35:04.837 INFO  ProgressMeter -      Reference:13723              0.2                   120            660.7\n",
      "11:35:15.163 INFO  ProgressMeter -      Reference:28507              0.4                   250            706.7\n",
      "11:35:25.957 INFO  ProgressMeter -      Reference:40232              0.5                   340            637.1\n",
      "11:35:36.294 INFO  ProgressMeter -      Reference:54810              0.7                   470            665.8\n",
      "11:35:47.229 INFO  ProgressMeter -      Reference:71365              0.9                   620            698.1\n",
      "11:35:57.277 INFO  ProgressMeter -      Reference:85966              1.1                   750            710.5\n",
      "11:36:07.387 INFO  ProgressMeter -      Reference:96881              1.2                   840            686.2\n",
      "11:36:17.699 INFO  ProgressMeter -     Reference:110339              1.4                   950            680.5\n",
      "11:36:28.544 INFO  ProgressMeter -     Reference:126041              1.6                  1090            691.3\n",
      "11:36:39.714 INFO  ProgressMeter -     Reference:141143              1.8                  1220            692.0\n",
      "11:36:49.837 INFO  ProgressMeter -     Reference:154260              1.9                  1330            688.5\n",
      "11:37:00.545 INFO  ProgressMeter -     Reference:166500              2.1                  1430            677.7\n",
      "11:37:11.929 INFO  ProgressMeter -     Reference:184972              2.3                  1600            695.7\n",
      "11:37:22.621 INFO  ProgressMeter -     Reference:203602              2.5                  1750            706.2\n",
      "11:37:32.839 INFO  ProgressMeter -     Reference:216263              2.6                  1860            702.3\n",
      "11:37:43.346 INFO  ProgressMeter -     Reference:229959              2.8                  1980            701.3\n",
      "11:37:53.376 INFO  ProgressMeter -     Reference:242365              3.0                  2080            695.5\n",
      "11:38:04.256 INFO  ProgressMeter -     Reference:255886              3.2                  2200            693.6\n",
      "11:38:14.342 INFO  ProgressMeter -     Reference:267649              3.3                  2300            688.6\n",
      "11:38:24.917 INFO  ProgressMeter -     Reference:276566              3.5                  2370            674.0\n",
      "11:38:35.180 INFO  ProgressMeter -     Reference:291348              3.7                  2500            678.0\n",
      "11:38:47.536 INFO  ProgressMeter -     Reference:306633              3.9                  2620            673.0\n",
      "11:38:57.964 INFO  ProgressMeter -     Reference:365129              4.1                  2900            713.0\n",
      "11:39:08.340 INFO  ProgressMeter -     Reference:413554              4.2                  3150            742.9\n",
      "11:39:18.977 INFO  ProgressMeter -     Reference:423927              4.4                  3240            733.5\n",
      "11:39:29.718 INFO  ProgressMeter -     Reference:436534              4.6                  3350            728.8\n",
      "11:39:40.104 INFO  ProgressMeter -     Reference:452302              4.8                  3490            731.7\n",
      "11:39:50.364 INFO  ProgressMeter -     Reference:468959              4.9                  3630            734.8\n",
      "11:40:00.935 INFO  ProgressMeter -     Reference:483391              5.1                  3750            732.9\n",
      "11:40:12.443 INFO  ProgressMeter -     Reference:498961              5.3                  3890            732.8\n",
      "11:40:22.482 INFO  ProgressMeter -     Reference:509649              5.5                  3980            726.8\n",
      "11:40:32.535 INFO  ProgressMeter -     Reference:522634              5.6                  4100            726.5\n",
      "11:40:43.121 INFO  ProgressMeter -     Reference:536328              5.8                  4210            723.4\n",
      "11:40:53.149 INFO  ProgressMeter -     Reference:552061              6.0                  4340            724.9\n",
      "11:41:03.492 INFO  ProgressMeter -     Reference:565632              6.2                  4470            725.7\n",
      "11:41:14.893 INFO  ProgressMeter -     Reference:577267              6.3                  4580            721.3\n",
      "11:41:25.298 INFO  ProgressMeter -     Reference:591199              6.5                  4700            720.6\n",
      "11:41:35.597 INFO  ProgressMeter -     Reference:605298              6.7                  4820            720.0\n",
      "11:41:46.926 INFO  ProgressMeter -     Reference:615941              6.9                  4910            713.3\n",
      "11:41:58.438 INFO  ProgressMeter -     Reference:627796              7.1                  5010            708.1\n",
      "11:42:08.749 INFO  ProgressMeter -     Reference:640787              7.2                  5120            706.5\n",
      "11:42:19.337 INFO  ProgressMeter -     Reference:655939              7.4                  5260            708.6\n",
      "11:42:29.789 INFO  ProgressMeter -     Reference:668343              7.6                  5360            705.5\n",
      "11:42:40.521 INFO  ProgressMeter -     Reference:683587              7.8                  5500            707.3\n",
      "11:42:52.150 INFO  ProgressMeter -     Reference:698834              8.0                  5630            706.4\n",
      "11:43:02.314 INFO  ProgressMeter -     Reference:712242              8.1                  5740            705.2\n",
      "11:43:13.382 INFO  ProgressMeter -     Reference:725736              8.3                  5860            704.0\n",
      "11:43:24.762 INFO  ProgressMeter -     Reference:735921              8.5                  5940            697.7\n",
      "11:43:35.049 INFO  ProgressMeter -     Reference:749726              8.7                  6050            696.6\n",
      "11:43:45.791 INFO  ProgressMeter -     Reference:764124              8.9                  6170            696.1\n",
      "11:43:57.258 INFO  ProgressMeter -     Reference:782557              9.1                  6340            700.1\n",
      "11:44:08.034 INFO  ProgressMeter -     Reference:797210              9.2                  6470            700.6\n",
      "11:44:18.188 INFO  ProgressMeter -     Reference:810664              9.4                  6580            699.7\n",
      "11:44:28.331 INFO  ProgressMeter -     Reference:826077              9.6                  6710            700.9\n",
      "11:44:39.412 INFO  ProgressMeter -     Reference:838345              9.8                  6810            697.9\n",
      "11:44:50.993 INFO  ProgressMeter -     Reference:849586             10.0                  6910            694.4\n",
      "11:45:01.029 INFO  ProgressMeter -     Reference:865930             10.1                  7040            695.8\n",
      "11:45:14.557 INFO  ProgressMeter -     Reference:884395             10.3                  7210            697.0\n",
      "11:45:25.609 INFO  ProgressMeter -     Reference:901021             10.5                  7350            698.1\n",
      "11:45:36.179 INFO  ProgressMeter -     Reference:915681             10.7                  7480            698.8\n",
      "11:45:46.616 INFO  ProgressMeter -     Reference:928640             10.9                  7600            698.7\n",
      "11:45:56.823 INFO  ProgressMeter -     Reference:946776             11.0                  7760            702.4\n",
      "11:46:07.149 INFO  ProgressMeter -     Reference:958862             11.2                  7860            700.5\n",
      "11:46:17.275 INFO  ProgressMeter -     Reference:974673             11.4                  8000            702.4\n",
      "11:46:27.634 INFO  ProgressMeter -     Reference:985796             11.6                  8090            699.7\n",
      "11:46:37.314 WARN  IntelInflater - Zero Bytes Written : 0\n",
      "11:46:37.507 INFO  HaplotypeCaller - 23848 read(s) filtered by: MappingQualityReadFilter \n",
      "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
      "0 read(s) filtered by: MappedReadFilter \n",
      "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
      "162 read(s) filtered by: NotDuplicateReadFilter \n",
      "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
      "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
      "0 read(s) filtered by: GoodCigarReadFilter \n",
      "0 read(s) filtered by: WellformedReadFilter \n",
      "24010 total reads filtered\n",
      "11:46:37.507 INFO  ProgressMeter -    Reference:1019520             11.7                  8281            706.2\n",
      "11:46:37.507 INFO  ProgressMeter - Traversal complete. Processed 8281 total regions in 11.7 minutes.\n",
      "11:46:37.525 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.273822774\n",
      "11:46:37.525 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 61.127270705\n",
      "11:46:37.525 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 549.45 sec\n",
      "11:46:37.525 INFO  HaplotypeCaller - Shutting down engine\n",
      "[June 17, 2022 at 11:46:37 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 11.74 minutes.\n",
      "Runtime.totalMemory()=772800512\n",
      "\n",
      "Calling Clone3\n",
      "Using GATK jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4g -jar /opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar HaplotypeCaller --native-pair-hmm-threads 4 -I Clone3.SORTED.MD.bam -O /home/jovyan/work/VCF/Clone3.g.vcf -R /home/jovyan/work/SV_DATA/REF/reference.fasta -ERC GVCF\n",
      "11:46:40.646 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "11:46:40.817 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "11:46:40.817 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.6.1\n",
      "11:46:40.817 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "11:46:40.822 INFO  HaplotypeCaller - Initializing engine\n",
      "11:46:41.013 INFO  HaplotypeCaller - Done initializing engine\n",
      "11:46:41.015 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
      "11:46:41.021 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
      "11:46:41.021 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
      "11:46:41.041 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
      "11:46:41.043 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/opt/conda/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
      "11:46:41.064 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
      "11:46:41.083 INFO  IntelPairHmm - Available threads: 8\n",
      "11:46:41.083 INFO  IntelPairHmm - Requested threads: 4\n",
      "11:46:41.084 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
      "11:46:41.163 INFO  ProgressMeter - Starting traversal\n",
      "11:46:41.164 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "11:46:41.825 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position Reference:54 and possibly subsequent; at least 10 samples must have called genotypes\n",
      "11:46:52.309 INFO  ProgressMeter -       Reference:5170              0.2                    30            161.5\n",
      "11:47:03.577 INFO  ProgressMeter -      Reference:11919              0.4                    70            187.4\n",
      "11:47:17.459 INFO  ProgressMeter -      Reference:19220              0.6                   110            181.8\n",
      "11:47:28.133 INFO  ProgressMeter -      Reference:26396              0.8                   150            191.6\n",
      "11:47:40.683 INFO  ProgressMeter -      Reference:33254              1.0                   190            191.5\n",
      "11:47:51.715 INFO  ProgressMeter -      Reference:39813              1.2                   230            195.6\n",
      "11:48:03.420 INFO  ProgressMeter -      Reference:46084              1.4                   270            196.9\n",
      "11:48:14.732 INFO  ProgressMeter -      Reference:55897              1.6                   330            211.6\n",
      "11:48:25.391 INFO  ProgressMeter -      Reference:64344              1.7                   380            218.8\n",
      "11:48:37.024 INFO  ProgressMeter -      Reference:72575              1.9                   430            222.7\n",
      "11:48:48.139 INFO  ProgressMeter -      Reference:81751              2.1                   480            226.8\n",
      "11:48:58.333 INFO  ProgressMeter -      Reference:89541              2.3                   530            231.8\n",
      "11:49:09.443 INFO  ProgressMeter -      Reference:98034              2.5                   580            234.7\n",
      "11:49:21.896 INFO  ProgressMeter -     Reference:104451              2.7                   620            231.4\n",
      "11:49:36.777 INFO  ProgressMeter -     Reference:111438              2.9                   660            225.5\n",
      "11:49:48.125 INFO  ProgressMeter -     Reference:118776              3.1                   700            224.6\n",
      "11:49:58.844 INFO  ProgressMeter -     Reference:123762              3.3                   730            221.6\n",
      "11:50:10.292 INFO  ProgressMeter -     Reference:133166              3.5                   790            226.7\n",
      "11:50:23.998 INFO  ProgressMeter -     Reference:139135              3.7                   820            220.8\n",
      "11:50:36.616 INFO  ProgressMeter -     Reference:144326              3.9                   850            216.6\n",
      "11:50:47.679 INFO  ProgressMeter -     Reference:151057              4.1                   890            216.6\n",
      "11:51:01.641 INFO  ProgressMeter -     Reference:157972              4.3                   930            214.2\n",
      "11:51:12.467 INFO  ProgressMeter -     Reference:164399              4.5                   970            214.5\n",
      "11:51:24.009 INFO  ProgressMeter -     Reference:172589              4.7                  1020            216.4\n",
      "11:51:35.745 INFO  ProgressMeter -     Reference:181135              4.9                  1070            217.9\n",
      "11:51:46.232 INFO  ProgressMeter -     Reference:189982              5.1                  1130            222.2\n",
      "11:51:58.561 INFO  ProgressMeter -     Reference:198448              5.3                  1180            223.1\n",
      "11:52:11.718 INFO  ProgressMeter -     Reference:206734              5.5                  1230            223.3\n",
      "11:52:22.386 INFO  ProgressMeter -     Reference:213180              5.7                  1270            223.3\n",
      "11:52:32.493 INFO  ProgressMeter -     Reference:220312              5.9                  1310            223.7\n",
      "11:52:45.520 INFO  ProgressMeter -     Reference:229008              6.1                  1360            224.0\n",
      "11:52:55.917 INFO  ProgressMeter -     Reference:235382              6.2                  1400            224.1\n",
      "11:53:08.312 INFO  ProgressMeter -     Reference:244356              6.5                  1450            224.7\n",
      "11:53:22.173 INFO  ProgressMeter -     Reference:252331              6.7                  1500            224.4\n",
      "11:53:33.978 INFO  ProgressMeter -     Reference:260458              6.9                  1550            225.3\n",
      "11:53:44.267 INFO  ProgressMeter -     Reference:267567              7.1                  1590            225.5\n",
      "11:53:54.580 INFO  ProgressMeter -     Reference:274650              7.2                  1630            225.6\n",
      "11:54:07.282 INFO  ProgressMeter -     Reference:280672              7.4                  1670            224.6\n",
      "11:54:19.181 INFO  ProgressMeter -     Reference:287585              7.6                  1710            224.0\n",
      "11:54:30.855 INFO  ProgressMeter -     Reference:297588              7.8                  1770            226.1\n",
      "11:54:42.746 INFO  ProgressMeter -     Reference:307593              8.0                  1840            229.2\n",
      "11:54:55.832 INFO  ProgressMeter -     Reference:315359              8.2                  1890            229.2\n",
      "11:55:06.979 INFO  ProgressMeter -     Reference:366388              8.4                  2090            247.9\n",
      "11:55:17.319 INFO  ProgressMeter -     Reference:405292              8.6                  2250            261.5\n",
      "11:55:28.889 INFO  ProgressMeter -     Reference:413929              8.8                  2300            261.5\n",
      "11:55:40.936 INFO  ProgressMeter -     Reference:420603              9.0                  2340            260.1\n",
      "11:55:53.303 INFO  ProgressMeter -     Reference:426503              9.2                  2370            257.5\n",
      "11:56:04.694 INFO  ProgressMeter -     Reference:435254              9.4                  2420            257.7\n",
      "11:56:16.778 INFO  ProgressMeter -     Reference:443685              9.6                  2470            257.5\n",
      "11:56:27.616 INFO  ProgressMeter -     Reference:450020              9.8                  2510            256.8\n",
      "11:56:39.593 INFO  ProgressMeter -     Reference:459063             10.0                  2560            256.7\n",
      "11:56:55.034 INFO  ProgressMeter -     Reference:465913             10.2                  2600            254.1\n",
      "11:57:05.397 INFO  ProgressMeter -     Reference:473033             10.4                  2640            253.8\n",
      "11:57:18.609 INFO  ProgressMeter -     Reference:481452             10.6                  2690            253.2\n",
      "11:57:31.465 INFO  ProgressMeter -     Reference:488041             10.8                  2730            251.9\n",
      "11:57:41.606 INFO  ProgressMeter -     Reference:495820             11.0                  2780            252.6\n",
      "11:57:51.852 INFO  ProgressMeter -     Reference:502933             11.2                  2820            252.3\n",
      "11:58:03.107 INFO  ProgressMeter -     Reference:510670             11.4                  2870            252.5\n",
      "11:58:15.129 INFO  ProgressMeter -     Reference:519380             11.6                  2920            252.5\n",
      "11:58:27.104 INFO  ProgressMeter -     Reference:527154             11.8                  2970            252.4\n",
      "11:58:38.413 INFO  ProgressMeter -     Reference:534750             12.0                  3010            251.8\n",
      "11:58:48.738 INFO  ProgressMeter -     Reference:542508             12.1                  3060            252.3\n",
      "11:59:00.246 INFO  ProgressMeter -     Reference:553953             12.3                  3120            253.3\n",
      "11:59:12.219 INFO  ProgressMeter -     Reference:562459             12.5                  3170            253.2\n",
      "11:59:25.549 INFO  ProgressMeter -     Reference:570303             12.7                  3220            252.8\n",
      "11:59:37.667 INFO  ProgressMeter -     Reference:576194             12.9                  3250            251.1\n",
      "11:59:47.736 INFO  ProgressMeter -     Reference:584287             13.1                  3300            251.7\n",
      "11:59:58.577 INFO  ProgressMeter -     Reference:590131             13.3                  3330            250.6\n",
      "12:00:09.934 INFO  ProgressMeter -     Reference:597146             13.5                  3370            250.0\n",
      "12:00:21.052 INFO  ProgressMeter -     Reference:603780             13.7                  3410            249.5\n",
      "12:00:32.628 INFO  ProgressMeter -     Reference:610213             13.9                  3450            249.0\n",
      "12:00:43.350 INFO  ProgressMeter -     Reference:615423             14.0                  3480            247.9\n",
      "12:00:53.515 INFO  ProgressMeter -     Reference:620274             14.2                  3510            247.1\n",
      "12:01:05.269 INFO  ProgressMeter -     Reference:626940             14.4                  3550            246.5\n",
      "12:01:15.411 INFO  ProgressMeter -     Reference:632356             14.6                  3580            245.7\n",
      "12:01:26.817 INFO  ProgressMeter -     Reference:639288             14.8                  3620            245.2\n",
      "12:01:37.289 INFO  ProgressMeter -     Reference:644814             14.9                  3650            244.4\n",
      "12:01:49.134 INFO  ProgressMeter -     Reference:650186             15.1                  3680            243.2\n",
      "12:02:01.954 INFO  ProgressMeter -     Reference:658472             15.3                  3730            243.1\n",
      "12:02:13.322 INFO  ProgressMeter -     Reference:663131             15.5                  3760            242.0\n",
      "12:02:25.388 INFO  ProgressMeter -     Reference:670623             15.7                  3800            241.5\n",
      "12:02:36.773 INFO  ProgressMeter -     Reference:678165             15.9                  3850            241.7\n",
      "12:02:51.122 INFO  ProgressMeter -     Reference:687136             16.2                  3900            241.2\n",
      "12:03:03.516 INFO  ProgressMeter -     Reference:693720             16.4                  3940            240.6\n",
      "12:03:15.845 INFO  ProgressMeter -     Reference:701943             16.6                  3990            240.7\n",
      "12:03:27.640 INFO  ProgressMeter -     Reference:711879             16.8                  4050            241.4\n",
      "12:03:38.709 INFO  ProgressMeter -     Reference:718706             17.0                  4090            241.2\n",
      "12:03:49.605 INFO  ProgressMeter -     Reference:727731             17.1                  4150            242.1\n",
      "12:04:04.364 INFO  ProgressMeter -     Reference:735409             17.4                  4190            241.0\n",
      "12:04:14.676 INFO  ProgressMeter -     Reference:740934             17.6                  4220            240.3\n",
      "12:04:29.519 INFO  ProgressMeter -     Reference:749690             17.8                  4270            239.8\n",
      "12:04:41.706 INFO  ProgressMeter -     Reference:756543             18.0                  4310            239.3\n",
      "12:04:53.341 INFO  ProgressMeter -     Reference:766246             18.2                  4370            240.1\n",
      "12:05:04.432 INFO  ProgressMeter -     Reference:775931             18.4                  4420            240.4\n",
      "12:05:15.668 INFO  ProgressMeter -     Reference:784625             18.6                  4470            240.6\n",
      "12:05:26.848 INFO  ProgressMeter -     Reference:792736             18.8                  4520            240.9\n",
      "12:05:37.585 INFO  ProgressMeter -     Reference:799993             18.9                  4560            240.8\n",
      "12:05:49.341 INFO  ProgressMeter -     Reference:806558             19.1                  4600            240.4\n",
      "12:06:01.616 INFO  ProgressMeter -     Reference:814094             19.3                  4640            239.9\n",
      "12:06:15.984 INFO  ProgressMeter -     Reference:823935             19.6                  4700            240.0\n",
      "12:06:28.046 INFO  ProgressMeter -     Reference:830375             19.8                  4740            239.6\n",
      "12:06:38.322 INFO  ProgressMeter -     Reference:838523             20.0                  4790            240.1\n",
      "12:06:52.596 INFO  ProgressMeter -     Reference:846901             20.2                  4840            239.7\n",
      "12:07:04.028 INFO  ProgressMeter -     Reference:853590             20.4                  4880            239.4\n",
      "12:07:18.316 INFO  ProgressMeter -     Reference:860792             20.6                  4920            238.6\n",
      "12:07:28.479 INFO  ProgressMeter -     Reference:867828             20.8                  4960            238.6\n",
      "12:07:40.449 INFO  ProgressMeter -     Reference:877743             21.0                  5020            239.2\n",
      "12:07:51.300 INFO  ProgressMeter -     Reference:886934             21.2                  5070            239.5\n",
      "12:08:02.559 INFO  ProgressMeter -     Reference:896183             21.4                  5120            239.7\n",
      "12:08:15.351 INFO  ProgressMeter -     Reference:902858             21.6                  5160            239.2\n",
      "12:08:26.452 INFO  ProgressMeter -     Reference:908909             21.8                  5200            239.0\n",
      "12:08:37.333 INFO  ProgressMeter -     Reference:916295             21.9                  5240            238.9\n",
      "12:08:50.422 INFO  ProgressMeter -     Reference:924918             22.2                  5290            238.8\n",
      "12:09:01.508 INFO  ProgressMeter -     Reference:933427             22.3                  5340            239.0\n"
     ]
    }
   ],
   "source": [
    "cd $BAM_ILL\n",
    "for i in {1..20}\n",
    "    do\n",
    "        # lauch GATK HaplotypeCaller\n",
    "        echo -e \"\\nCalling Clone$i\";\n",
    "        gatk --java-options \"-Xmx4g\" HaplotypeCaller --native-pair-hmm-threads 4 -I Clone$i.SORTED.MD.bam -O $VCF_DIR/Clone$i.g.vcf -R $REF -ERC GVCF\n",
    "    done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07a52085-ad7b-4833-b8c4-845085a2ff23",
   "metadata": {},
   "source": [
    "### CombineGVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6afa0781-11db-4c1a-ad29-5d8c4dbe1efd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# change of work directory\n",
    "cd $VCF_DIR\n",
    "# Loop to inflate the --variant option\n",
    "OPTION=\"\"\n",
    "for i in {1..20}\n",
    "do\n",
    "    OPTION=\"${OPTION} --variant Clone${i}.g.vcf\"\n",
    "done\n",
    "echo $OPTION\n",
    "# GATK\n",
    "gatk CombineGVCFs -R $REF $OPTION -O rawSNP.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d2f82d-dfc9-45ef-bca4-c186c356b4fe",
   "metadata": {},
   "source": [
    "### Have a Look to it combineGVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dbd61ee-0e7d-4bda-bc26-bfeb0830193f",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 1000 rawSNP.vcf | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ec071c-36de-498a-bc4b-2099bd1308e4",
   "metadata": {},
   "source": [
    "### Compute the Genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6fbf4fe-d12f-4bf3-92b9-9aa446867e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "gatk --java-options \"-Xmx4g\" GenotypeGVCFs -R $REF -V rawSNP.vcf -O output.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf02a88d-4832-417e-92bf-f1a994976844",
   "metadata": {},
   "source": [
    "### Compute the SNP density along the chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57505229-54d7-42b0-b470-4be5b11e420c",
   "metadata": {},
   "outputs": [],
   "source": [
    "echo -e \"Reference\\t1000000\\n\" > genome.txt\n",
    "bedtools genomecov -bga -split -i output.vcf -g $REF > density.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a00fc838-fd28-4f12-9871-c1b4c89cdcbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "head density.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d27e2f39-7f7b-48e5-bd26-d1bd17b6b1b1",
   "metadata": {},
   "source": [
    "### Count the number of variants with `bcftools stat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8665135f-a849-4711-bafe-9c8ebe54f250",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools stats output.vcf > output.vcf.stats"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
