{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "22cccec7-858e-4e84-b967-feb1b5f78a58",
   "metadata": {},
   "source": [
    "# Get SNP file and calculate some statistics\n",
    "\n",
    "This tutorial is based on a SNP dataset extracted from the 3000 Rice genome project. For limiting time of computation, we restricted the dataset to chr1 for 60 individuals (30 indica and 30 japonica)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6cc1899-bc1b-4fea-be70-e99215262875",
   "metadata": {},
   "source": [
    "#### Create a new directory 'rice3k' for this analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "32811024-37fe-4ba5-8e10-14ab39258995",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/home/jovyan/rice3k’: File exists\n"
     ]
    }
   ],
   "source": [
    "mkdir /home/jovyan/rice3k\n",
    "cd /home/jovyan/rice3k"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5ec4d29-ba38-4169-a6c7-85a357f3165e",
   "metadata": {},
   "source": [
    "#### Download the VCF dataset\n",
    "\n",
    "The wget command allows to download any dataset available on the web, through a URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f064e29b-09e7-4f52-9457-b49a16603947",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-09-07 13:28:13--  https://itrop.ird.fr/sv-training/out.vcf.gz\n",
      "Resolving itrop.ird.fr (itrop.ird.fr)... 91.203.35.184\n",
      "Connecting to itrop.ird.fr (itrop.ird.fr)|91.203.35.184|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1678691 (1.6M) [application/x-gzip]\n",
      "Saving to: ‘out.vcf.gz’\n",
      "\n",
      "out.vcf.gz          100%[===================>]   1.60M  --.-KB/s    in 0.1s    \n",
      "\n",
      "2022-09-07 13:28:14 (15.0 MB/s) - ‘out.vcf.gz’ saved [1678691/1678691]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wget https://itrop.ird.fr/sv-training/out.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a54074-b723-445c-a0a2-d1eff1d6e5ce",
   "metadata": {},
   "source": [
    "#### Some statistics about SNPs\n",
    "\n",
    "Count the number of variants with `bcftools stat`\n",
    "- Run the bcftools stats on the vcf file and save the result into the file `SNP_statistics.txt`\n",
    "- How many samples were used for this SNP analysis ?\n",
    "- How many SNPs were detected ? Is there any other easy way to identify the number of variants in VCF file?\n",
    "- What is the ratio transition/transversion?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87c53599-84ff-45e5-a864-17e525195f47",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools stats out.vcf.gz >SNP_statistics.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63693f73-e998-4760-b16e-bcfe54f193c9",
   "metadata": {},
   "source": [
    "#### Calculate allele frequency\n",
    "\n",
    "Calculate allele frequency of each position using `vcftools`\n",
    "\n",
    "--freq2 : outputs the frequencies without information about the alleles\n",
    "\n",
    "--freq would return their identity\n",
    "\n",
    "--max-alleles 2 to exclude sites that have more than two alleles.\n",
    "\n",
    "Compare outputs between these two options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "385b8d15-74fb-43e5-a028-be94c1df3e74",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--max-alleles 2\n",
      "\t--freq\n",
      "\t--out AF\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "After filtering, kept 60 out of 60 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 119202 out of a possible 119202 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcftools --gzvcf out.vcf.gz --freq --out AF --max-alleles 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "08048370-ec73-445d-81ae-73aaecc27736",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--max-alleles 2\n",
      "\t--freq2\n",
      "\t--out AF2\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "After filtering, kept 60 out of 60 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 119202 out of a possible 119202 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcftools --gzvcf out.vcf.gz --freq2 --out AF2 --max-alleles 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3da2123-57ce-4f64-978c-e1063a63d03c",
   "metadata": {},
   "source": [
    "# Annotate your SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e56bef7e-fa28-4fd6-b8ee-216d065e1efe",
   "metadata": {},
   "source": [
    "### Install and configure snpEff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f0353eb-6417-4485-88f9-d2e0a45551eb",
   "metadata": {},
   "source": [
    "We will use `snpEff` software to perform SNP annotation. \n",
    "\n",
    "SNP annotation consists of predicting the effects of genetic variants on genes and proteins (such as amino acid changes).\n",
    "\n",
    "First, download and unzip the snpEff archive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0e3df5a-1405-47a0-b305-1da1c5b4ca8a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-09-07 14:33:39--  https://netix.dl.sourceforge.net/project/snpeff/snpEff_v4_3t_core.zip\n",
      "Resolving netix.dl.sourceforge.net (netix.dl.sourceforge.net)... 87.121.121.2\n",
      "Connecting to netix.dl.sourceforge.net (netix.dl.sourceforge.net)|87.121.121.2|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 69708122 (66M) [application/octet-stream]\n",
      "Saving to: ‘snpEff_v4_3t_core.zip’\n",
      "\n",
      "snpEff_v4_3t_core.z 100%[===================>]  66.48M  40.1MB/s    in 1.7s    \n",
      "\n",
      "2022-09-07 14:33:41 (40.1 MB/s) - ‘snpEff_v4_3t_core.zip’ saved [69708122/69708122]\n",
      "\n",
      "Archive:  snpEff_v4_3t_core.zip\n",
      "   creating: snpEff/\n",
      "   creating: snpEff/galaxy/\n",
      "  inflating: snpEff/galaxy/snpEff.xml  \n",
      "  inflating: snpEff/galaxy/snpSift_caseControl.xml  \n",
      "  inflating: snpEff/galaxy/snpEff_download.xml  \n",
      "  inflating: snpEff/galaxy/snpEffWrapper.pl  \n",
      "  inflating: snpEff/galaxy/snpSift_filter.xml  \n",
      "  inflating: snpEff/galaxy/snpSift_int.xml  \n",
      "  inflating: snpEff/galaxy/snpSift_annotate.xml  \n",
      "  inflating: snpEff/galaxy/snpSiftWrapper.pl  \n",
      "  inflating: snpEff/galaxy/tool_dependencies.xml  \n",
      "   creating: snpEff/galaxy/tool-data/\n",
      "  inflating: snpEff/galaxy/tool-data/snpEff_genomes.loc  \n",
      "  inflating: snpEff/galaxy/tool-data/snpEff_genomes.loc.sample  \n",
      "  inflating: snpEff/galaxy/tool_conf.xml  \n",
      "   creating: snpEff/scripts/\n",
      "  inflating: snpEff/scripts/countColumns.py  \n",
      "  inflating: snpEff/scripts/vcfAnnFirst.py  \n",
      "  inflating: snpEff/scripts/join.pl  \n",
      "  inflating: snpEff/scripts/txt2vcf.py  \n",
      "  inflating: snpEff/scripts/fastqSplit.pl  \n",
      "  inflating: snpEff/scripts/cgShore.sh  \n",
      "  inflating: snpEff/scripts/annotate_demo_GATK.sh  \n",
      "  inflating: snpEff/scripts/vcfInfoOnePerLine.pl  \n",
      "  inflating: snpEff/scripts/fasta2tab.pl  \n",
      "  inflating: snpEff/scripts/vcfOnlyAlts.pl  \n",
      "  inflating: snpEff/scripts/ped2vcf.py  \n",
      "  inflating: snpEff/scripts/vcfReduceGenotypes.pl  \n",
      "  inflating: snpEff/scripts/transpose.pl  \n",
      "  inflating: snpEff/scripts/1kg.sh   \n",
      "  inflating: snpEff/scripts/splitChr.pl  \n",
      "  inflating: snpEff/scripts/snpSift_filter_sample_to_number.pl  \n",
      "  inflating: snpEff/scripts/plotLabel.pl  \n",
      "  inflating: snpEff/scripts/filterBy.py  \n",
      "  inflating: snpEff/scripts/plotMA.pl  \n",
      "  inflating: snpEff/scripts/sam2fastq.pl  \n",
      "  inflating: snpEff/scripts/vcfEffHighest.ORI.py  \n",
      "  inflating: snpEff/scripts/swapCols.pl  \n",
      "  inflating: snpEff/scripts/wigSplit.pl  \n",
      "  inflating: snpEff/scripts/queue.pl  \n",
      "  inflating: snpEff/scripts/nextProt_filter.pl  \n",
      "  inflating: snpEff/scripts/uniqCount.pl  \n",
      "  inflating: snpEff/scripts/snpEff   \n",
      "  inflating: snpEff/scripts/vcfBareBones.pl  \n",
      "  inflating: snpEff/scripts/plotXY.pl  \n",
      "  inflating: snpEff/scripts/plotQQsubsample.pl  \n",
      "  inflating: snpEff/scripts/joinSnpEff.pl  \n",
      "  inflating: snpEff/scripts/gffRemovePhase.pl  \n",
      "  inflating: snpEff/scripts/fastaSplit.pl  \n",
      "  inflating: snpEff/scripts/buildDbNcbi.sh  \n",
      "   creating: snpEff/scripts/gsa/\n",
      "  inflating: snpEff/scripts/gsa/pvalue_correction_scoreCount_min10.sh  \n",
      "  inflating: snpEff/scripts/gsa/bayesFactor_correction_scoreCount.sh  \n",
      "  inflating: snpEff/scripts/gsa/geneSetOverlap.sort.txt  \n",
      "  inflating: snpEff/scripts/gsa/pvalue_correction_scoreCount.r  \n",
      "  inflating: snpEff/scripts/gsa/bayesFactor_correction_scoreCount_max10.sh  \n",
      "  inflating: snpEff/scripts/gsa/geneSetOverlap.py  \n",
      "  inflating: snpEff/scripts/gsa/pvalue_correction_scoreCount.sh  \n",
      "  inflating: snpEff/scripts/gsa/create_sets.bds  \n",
      "  inflating: snpEff/scripts/gsa/checkGeneNames.py  \n",
      "  inflating: snpEff/scripts/gsa/geneSetsGtex.py  \n",
      "  inflating: snpEff/scripts/gsa/bayesFactor_correction_scoreCount.r  \n",
      "  inflating: snpEff/scripts/uniqCut.pl  \n",
      "  inflating: snpEff/scripts/plot.pl  \n",
      "  inflating: snpEff/scripts/cgShore.pl  \n",
      "  inflating: snpEff/scripts/plotQQ.pl  \n",
      "  inflating: snpEff/scripts/fastaSample.pl  \n",
      "  inflating: snpEff/scripts/vcfFilterSamples.pl  \n",
      "  inflating: snpEff/scripts/isutf8.py  \n",
      "  inflating: snpEff/scripts/db.pl    \n",
      "  inflating: snpEff/scripts/vcfEffOnePerLine.pl  \n",
      "  inflating: snpEff/scripts/plotSmoothScatter.pl  \n",
      "  inflating: snpEff/scripts/make_dbNSFP.sh  \n",
      "  inflating: snpEff/scripts/bedEffOnePerLine.pl  \n",
      "  inflating: snpEff/scripts/plotHistogram.pl  \n",
      "  inflating: snpEff/scripts/sortLine.py  \n",
      "  inflating: snpEff/scripts/annotate_demo.sh  \n",
      "  inflating: snpEff/scripts/txt2fa.pl  \n",
      "  inflating: snpEff/scripts/statsNum.pl  \n",
      "  inflating: snpEff/scripts/vcfRefCorrect.py  \n",
      "  inflating: snpEff/scripts/extractSequences.pl  \n",
      "  inflating: snpEff/snpEff.config    \n",
      "  inflating: snpEff/SnpSift.jar      \n",
      "  inflating: snpEff/snpEff.jar       \n",
      "   creating: snpEff/examples/\n",
      "  inflating: snpEff/examples/test.chr22.vcf  \n",
      "  inflating: snpEff/examples/samples_cancer.txt  \n",
      "  inflating: snpEff/examples/samples_cancer_one.txt  \n",
      "  inflating: snpEff/examples/cancer_pedigree.vcf  \n",
      "  inflating: snpEff/examples/test.chr22.ann.filter_missense.vcf  \n",
      "  inflating: snpEff/examples/variants_1.ann.vcf  \n",
      "  inflating: snpEff/examples/cancer.vcf  \n",
      "  inflating: snpEff/examples/test.1KG.vcf  \n",
      "  inflating: snpEff/examples/intervals.bed  \n",
      "  inflating: snpEff/examples/test.chr22.ann.one_per_line.txt  \n",
      "  inflating: snpEff/examples/variants_2.vcf  \n",
      "  inflating: snpEff/examples/cancer.eff.vcf  \n",
      "  inflating: snpEff/examples/1kg.head_chr1.filtered.vcf.gz  \n",
      "  inflating: snpEff/examples/test.chr22.ann.filter_missense_any.vcf  \n",
      "  inflating: snpEff/examples/1kg.head_chr1.vcf.gz  \n",
      "  inflating: snpEff/examples/test.chr22.ann.vcf  \n",
      "  inflating: snpEff/examples/test.ann.vcf  \n",
      "  inflating: snpEff/examples/examples.sh  \n",
      "  inflating: snpEff/examples/my_annotations.bed  \n",
      " extracting: snpEff/examples/test.vcf  \n",
      "  inflating: snpEff/examples/test.chr22.ann.filter_missense_first.vcf  \n",
      "  inflating: snpEff/examples/test.chr22.ann.txt  \n",
      "  inflating: snpEff/examples/test.chr22.ann.filter_missense_any_TRMT2A.vcf  \n",
      "  inflating: snpEff/examples/cancer.ann.vcf  \n",
      "  inflating: snpEff/examples/variants_1.vcf  \n",
      "  inflating: snpEff/examples/example_motif.vcf  \n",
      "  inflating: snpEff/examples/cancer_pedigree.ann.vcf  \n",
      "  inflating: snpEff/examples/test.1KG.ann_encode.vcf  \n",
      "  inflating: snpEff/examples/file.vcf  \n",
      "  inflating: snpEff/examples/variants_2.ann.vcf  \n",
      "  inflating: snpEff/examples/test.1KG.ann_reg.vcf  \n",
      "replace clinEff/clinEff.config? [y]es, [n]o, [A]ll, [N]one, [r]ename: "
     ]
    }
   ],
   "source": [
    "wget https://netix.dl.sourceforge.net/project/snpeff/snpEff_v4_3t_core.zip\n",
    "unzip snpEff_v4_3t_core.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bcd86e9-2367-442c-b4a6-dfa05554b439",
   "metadata": {},
   "source": [
    "#### Look at the different genomes available for Rice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "557364e1-31ac-4b0e-bdaa-e704a1aafe2e",
   "metadata": {},
   "source": [
    "In order to annotate a VCF file, the program needs to know the location of genes in the genome on which the SNP are positionned.\n",
    "\n",
    "First, we will have to configure a new database corresponding to Oryza sativa Nipponbare (used as reference to generate the VCF file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d4d0c59f-3502-4763-9c3b-4d45420596ca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_barthii                                               \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_barthii                                               \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_brachyantha                                           \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_brachyantha                                           \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_glaberrima                                            \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_glaberrima                                            \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_glumaepatula                                          \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_glumaepatula                                          \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_indica                                                \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_indica                                                \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_longistaminata                                        \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_longistaminata                                        \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_meridionalis                                          \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_meridionalis                                          \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_nivara                                                \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_nivara                                                \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_punctata                                              \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_punctata                                              \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_rufipogon                                             \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_rufipogon                                             \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n",
      "\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_sativa                                                \t\u001b[01;31m\u001b[KOryza\u001b[m\u001b[K_sativa                                                \t          \tENSEMBL_BFMPP_32_268          \thttp://downloads.sourceforge.net/project/snpeff/databases/v4_3/snpEff_v4_3_ENSEMBL_BFMPP_32_268.zip\n"
     ]
    }
   ],
   "source": [
    "java -jar snpEff/snpEff.jar databases | grep Oryza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08fcc877-8ad7-44b3-93ee-8780fbf48b25",
   "metadata": {},
   "source": [
    "#### Download database for Rice (oryza_sativa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f67972-2ace-4d4a-99db-d227003ac425",
   "metadata": {},
   "source": [
    "The download of a new genome and its annotation may take a while"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "930ce26f-5aa9-42a6-966c-84103d85bd2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "java -jar snpEff/snpEff.jar download Oryza_sativa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7e89695-19de-4582-b5d8-2d1bb91abe03",
   "metadata": {},
   "source": [
    "Let's have a look to the files that have been retrieved. What are they?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc81336f-5379-4d59-b64e-bdfbb499a6d6",
   "metadata": {},
   "source": [
    "#### Annotate the VCF variant file (using the Rice genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d7772c1-7095-453c-8e66-f0394412f91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "java -jar snpEff/snpEff.jar eff Oryza_sativa out.vcf.gz >out.annotated.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48b7d104-28c0-4633-9377-14fd397ae3da",
   "metadata": {},
   "source": [
    "Look the new annotated VCF file, to see how the variants were annotated."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9b30fce-9500-4041-8f3e-a2cd92c540e9",
   "metadata": {},
   "source": [
    "#### Look at the resulting HTML snpEff output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe14f0bb-13bc-4b56-8fc3-1fc647d45241",
   "metadata": {},
   "source": [
    "<img src=\"snpeff_output.png\" align=\"center\" width=\"70%\" style=\"display: block; margin: auto;\"/> "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34f7f1f-e20a-43a6-9049-697db68bba50",
   "metadata": {},
   "source": [
    "# Play with SNPs and genotyping information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bcee8b0-5ad2-4fb7-8fed-534f758c985f",
   "metadata": {},
   "source": [
    "#### Install Plink software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "92f8c8fd-459d-4003-b2e6-2e1adc963839",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-09-07 15:21:30--  https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20220402.zip\n",
      "Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.36.102\n",
      "Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.36.102|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 8948563 (8.5M) [application/zip]\n",
      "Saving to: ‘plink_linux_x86_64_20220402.zip’\n",
      "\n",
      "plink_linux_x86_64_ 100%[===================>]   8.53M  9.86MB/s    in 0.9s    \n",
      "\n",
      "2022-09-07 15:21:32 (9.86 MB/s) - ‘plink_linux_x86_64_20220402.zip’ saved [8948563/8948563]\n",
      "\n",
      "Archive:  plink_linux_x86_64_20220402.zip\n",
      "  inflating: plink                   \n",
      "  inflating: LICENSE                 \n",
      "  inflating: toy.ped                 \n",
      "  inflating: toy.map                 \n",
      "  inflating: prettify                \n"
     ]
    }
   ],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20220402.zip\n",
    "unzip plink_linux_x86_64_20220402.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "166a31b8-3d7f-4a59-b67c-b4521b7de5c6",
   "metadata": {},
   "source": [
    "#### Evaluate level of missing data (by sample, by positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "76dde4d8-b5ef-470a-a0b5-315b3ece8e09",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.26 64-bit (2 Apr 2022)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to plink.log.\n",
      "Options in effect:\n",
      "  --missing\n",
      "  --vcf out.vcf.gz\n",
      "\n",
      "3933 MB RAM detected; reserving 1966 MB for main workspace.\n",
      "--vcf: plink-temporary.bed + plink-temporary.bim + plink-temporary.fam written.\n",
      "119202 variants loaded from .bim file.\n",
      "60 people (0 males, 0 females, 60 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to plink.nosex .\n",
      "Using 1 thread.\n",
      "Before main variant filters, 60 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.986788.\n",
      "--missing: Sample missing data report written to plink.imiss, and variant-based\n",
      "missing data report written to plink.lmiss.\n"
     ]
    }
   ],
   "source": [
    "./plink -vcf out.vcf.gz --missing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c26eb97-d3cd-4876-b5bc-6ff34b18e1fe",
   "metadata": {},
   "source": [
    "##### Plink generates 2 files. What information do they contain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "813698a2-8dc5-48de-95cd-f6ca3e7c8b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk -F \" \" {'print $1\"\\t\"$4'} plink.imiss | sort -n -k2 >missing.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "99670e80-4bbf-43fd-ad66-962b2a43a6bd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             FID              IID MISS_PHENO   N_MISS   N_GENO   F_MISS\n",
      "            B009             B009          Y      380   119202 0.003188\n",
      "            B010             B010          Y      385   119202  0.00323\n",
      "            B015             B015          Y       47   119202 0.0003943\n",
      "            B024             B024          Y      344   119202 0.002886\n",
      "            B059             B059          Y      914   119202 0.007668\n",
      "            B060             B060          Y      811   119202 0.006804\n",
      "            B061             B061          Y      431   119202 0.003616\n",
      "            B062             B062          Y     1250   119202  0.01049\n",
      "            B063             B063          Y      183   119202 0.001535\n"
     ]
    }
   ],
   "source": [
    "head plink.imiss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a371ed0-bb2d-4f1f-b3de-ba3eb6300e7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR        SNP   N_MISS   N_GENO   F_MISS\n",
      "   1       1579        0       60        0\n",
      "   1       1695        0       60        0\n",
      "   1       1708        0       60        0\n",
      "   1       1748        1       60  0.01667\n",
      "   1       1794        1       60  0.01667\n",
      "   1       1872        0       60        0\n",
      "   1       1948        1       60  0.01667\n",
      "   1       1953        1       60  0.01667\n",
      "   1       2065        1       60  0.01667\n"
     ]
    }
   ],
   "source": [
    "head plink.lmiss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b82256ba-75e2-4cc6-b1fc-d854494d8605",
   "metadata": {},
   "source": [
    "#### Distribution of missing levels across positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "894790ac-2699-4a85-a98e-ea5e9d619ac8",
   "metadata": {},
   "source": [
    "Using a `awk` command, very usefull for filtering tabulated file, we will extract positions (column 2) and the number of missing individuals for each position (column 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "68dcde17-7231-4c1f-9af8-be6a42ce02ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk -F \" \" {'print $1\"_\"$2\"\\t\"$3'} plink.lmiss >missing_by_snp.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "335c96c1-e56b-4a75-ad0a-0e663d3a192e",
   "metadata": {},
   "source": [
    "A short python script, using the pandas module, enables to display the distribution of missing data as an histogram. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "13c66d80-f924-4fdf-a0e3-cd2dd8bc65a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'N_MISS'}>]], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEICAYAAABfz4NwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVE0lEQVR4nO3dbYyd9Znf8e8vdkpcJzyHkWWjNS3Wqjxok2VKqdBG3nW6TB4aeAGqI20wrVOriKSsynZl8qJRVrIUWrFsQQuqC1kbwi64ZFl7k7IbanaaTcVDTJZdYwjFDRQmeHEJhGAkCEOvvjj/SY+HsT1zZsbnePz9SEfnPtd9/+/5X4zx79wP5zhVhSRJ7+v3BCRJg8FAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCBJJnk/ycpKlXbXPJxk9wrhq4xZ31RYn2Z+kumqjST7f9fpLSZ5LciDJWJJ7u9adm+TbSV5L8pMkjyf55Jw1Kx2GgSB1LAau7WHcT4BPdL3+JPDaoTZOsg74HPDxqvogMAzs7NrkT4EHgSHgDOBfAz/tYV7SjBkIUsd/AH4ryckzHHcXcGXX6yuBOw+z/T8E/ryq/hdAVf1tVW0GSHI6cBbwn6vqZ+3xP6rquzOck9QTA0Hq2AWMAr81w3F/AnwsycktTH4F2H6Y7R8Brkzyb5MMJ1nUte7HwF7g60kuSzI0w7lIs2IgSP/fvwO+mOTDMxjzFp3TPP8MWAvsaLUpVdXXgS8ClwD/HdifZGNbV8CvAs8DNwL7knwnyaqZtyLNnIEgNVX1JPBNYOMMh95J51TRkU4XTfycu6vq48DJwL8CfifJJW3dWFV9oar+PvALwJvT2ac0FwwE6WBfBv4lsHwGY/4SWEbnQvC0z/dX1TtV9V+AvwHOm2L9i8DvT7VOmg8GgtSlqvYC99K5u2e6Ywr4p8Bn6gj/wEiSq5J8KsmHkrwvySeAc4FHk5yS5CtJzm7rTgf+BZ3rDtK8MxCk9/odYOkRt+pSVXuqas80Nv0p8CXgBTq3rP574Op2J9HPgJXAf2vbPQm8DVw1k7lIvYr/YpokCTxCkCQ1i4+8iXR8SvIrwANTrWufMpYWFE8ZSZKAY/gI4fTTT6+VK1f2NPbNN99k6dIZXTMcWPYyeBZKH2Avg2o2vTz++OOvVNWUH748ZgNh5cqV7Nq1q6exo6OjrF69em4n1Cf2MngWSh9gL4NqNr0k+d+HWudFZUkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJwDH9SeTZ2/+h1rtr4rX5PY05sGVkYH8WX1H8eIUiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJKAaQZCkueT7E7yRJJdrXZqkgeTPNueT+na/voke5M8k+SSrvoFbT97k9ycJK1+QpJ7W/3RJCvnuE9J0hHM5AjhV6vqI1U13F5vBHZW1SpgZ3tNknOAtcC5wAhwa5JFbcxtwAZgVXuMtPp64LWqOhu4Cbih95YkSb2YzSmjS4GtbXkrcFlX/Z6qeruqngP2AhcmWQacWFUPV1UBd04aM7Gv+4A1E0cPkqSjY7pfblfAt5MU8J+qajMwVFX7AKpqX5Iz2rbLgUe6xo612jtteXJ9YsyLbV/jSV4HTgNe6Z5Ekg10jjAYGhpidHR0mtM/2NASuO788Z7GDpoDBw70/N9h0CyUXhZKH2Avg2q+epluIFxcVS+1v/QfTPKDw2w71Tv7Okz9cGMOLnSCaDPA8PBwrV69+rCTPpRb7t7OjbsXxhe9bhlZSq//HQbN6OjoguhlofQB9jKo5quXaZ0yqqqX2vN+4H7gQuDldhqI9ry/bT4GnNk1fAXwUquvmKJ+0Jgki4GTgFdn3o4kqVdHDIQkS5N8aGIZ+HXgSWAHsK5ttg7Y3pZ3AGvbnUNn0bl4/Fg7vfRGkova9YErJ42Z2NflwEPtOoMk6SiZznmTIeD+do13MfCHVfVnSb4HbEuyHngBuAKgqvYk2QY8BYwD11TVu21fVwNbgCXAA+0BcAdwV5K9dI4M1s5Bb5KkGThiIFTVD4FfmqL+Y2DNIcZsAjZNUd8FnDdF/S1aoEiS+sNPKkuSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCYDF/Z6AZmf3j17nqo3f6vc05sSWkaX9noJ0XPMIQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSmmkHQpJFSf4qyTfb61OTPJjk2fZ8Ste21yfZm+SZJJd01S9IsrutuzlJWv2EJPe2+qNJVs5hj5KkaZjJEcK1wNNdrzcCO6tqFbCzvSbJOcBa4FxgBLg1yaI25jZgA7CqPUZafT3wWlWdDdwE3NBTN5Kknk0rEJKsAD4F3N5VvhTY2pa3Apd11e+pqrer6jlgL3BhkmXAiVX1cFUVcOekMRP7ug9YM3H0IEk6Oqb7XUa/B/w28KGu2lBV7QOoqn1Jzmj15cAjXduNtdo7bXlyfWLMi21f40leB04DXumeRJINdI4wGBoaYnR0dJrTP9jQErju/PGexg6ahdTLgQMHev6dDpKF0gfYy6Car16OGAhJPg3sr6rHk6yexj6nemdfh6kfbszBharNwGaA4eHhWr16OtN5r1vu3s6NuxfG9/pdd/74gully8hSev2dDpLR0dEF0QfYy6Car16m8zfJxcBnknwS+ABwYpKvAy8nWdaODpYB+9v2Y8CZXeNXAC+1+oop6t1jxpIsBk4CXu2xJ0lSD454DaGqrq+qFVW1ks7F4oeq6jeAHcC6ttk6YHtb3gGsbXcOnUXn4vFj7fTSG0kuatcHrpw0ZmJfl7ef8Z4jBEnS/JnNuYavAtuSrAdeAK4AqKo9SbYBTwHjwDVV9W4bczWwBVgCPNAeAHcAdyXZS+fIYO0s5iVJ6sGMAqGqRoHRtvxjYM0httsEbJqivgs4b4r6W7RAkST1h59UliQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRIwjUBI8oEkjyX56yR7knyl1U9N8mCSZ9vzKV1jrk+yN8kzSS7pql+QZHdbd3OStPoJSe5t9UeTrJyHXiVJhzGdI4S3gV+rql8CPgKMJLkI2AjsrKpVwM72miTnAGuBc4ER4NYki9q+bgM2AKvaY6TV1wOvVdXZwE3ADbNvTZI0E0cMhOo40F6+vz0KuBTY2upbgcva8qXAPVX1dlU9B+wFLkyyDDixqh6uqgLunDRmYl/3AWsmjh4kSUfHtK4hJFmU5AlgP/BgVT0KDFXVPoD2fEbbfDnwYtfwsVZb3pYn1w8aU1XjwOvAaT30I0nq0eLpbFRV7wIfSXIycH+S8w6z+VTv7Osw9cONOXjHyQY6p5wYGhpidHT0MNM4tKElcN354z2NHTQLqZcDBw70/DsdJAulD7CXQTVfvUwrECZU1U+SjNI59/9ykmVVta+dDtrfNhsDzuwatgJ4qdVXTFHvHjOWZDFwEvDqFD9/M7AZYHh4uFavXj2T6f/cLXdv58bdM2p9YF13/viC6WXLyFJ6/Z0OktHR0QXRB9jLoJqvXqZzl9GH25EBSZYAHwd+AOwA1rXN1gHb2/IOYG27c+gsOhePH2unld5IclG7PnDlpDET+7oceKhdZ5AkHSXTeWu5DNja7hR6H7Ctqr6Z5GFgW5L1wAvAFQBVtSfJNuApYBy4pp1yArga2AIsAR5oD4A7gLuS7KVzZLB2LpqTJE3fEQOhqv4G+OgU9R8Daw4xZhOwaYr6LuA91x+q6i1aoEiS+sNPKkuSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSc0RAyHJmUn+IsnTSfYkubbVT03yYJJn2/MpXWOuT7I3yTNJLumqX5Bkd1t3c5K0+glJ7m31R5OsnIdeJUmHMZ0jhHHguqr6B8BFwDVJzgE2AjurahWws72mrVsLnAuMALcmWdT2dRuwAVjVHiOtvh54rarOBm4CbpiD3iRJM3DEQKiqfVX1/bb8BvA0sBy4FNjaNtsKXNaWLwXuqaq3q+o5YC9wYZJlwIlV9XBVFXDnpDET+7oPWDNx9CBJOjoWz2Tjdirno8CjwFBV7YNOaCQ5o222HHika9hYq73TlifXJ8a82PY1nuR14DTglUk/fwOdIwyGhoYYHR2dyfR/bmgJXHf+eE9jB81C6uXAgQM9/04HyULpA+xlUM1XL9MOhCQfBL4B/GZV/fQwb+CnWlGHqR9uzMGFqs3AZoDh4eFavXr1EWY9tVvu3s6Nu2eUhQPruvPHF0wvW0aW0uvvdJCMjo4uiD7AXgbVfPUyrbuMkryfThjcXVV/3Movt9NAtOf9rT4GnNk1fAXwUquvmKJ+0Jgki4GTgFdn2owkqXfTucsowB3A01X1u12rdgDr2vI6YHtXfW27c+gsOhePH2unl95IclHb55WTxkzs63LgoXadQZJ0lEznXMPFwOeA3UmeaLUvAV8FtiVZD7wAXAFQVXuSbAOeonOH0jVV9W4bdzWwBVgCPNAe0Amcu5LspXNksHZ2bUmSZuqIgVBV32Xqc/wAaw4xZhOwaYr6LuC8Kepv0QJFktQfflJZkgQYCJKkxkCQJAEGgiSpWRifaNKCsPtHr3PVxm/1exqztmVkab+nIPXEIwRJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEjCNQEjytST7kzzZVTs1yYNJnm3Pp3Stuz7J3iTPJLmkq35Bkt1t3c1J0uonJLm31R9NsnKOe5QkTcN0jhC2ACOTahuBnVW1CtjZXpPkHGAtcG4bc2uSRW3MbcAGYFV7TOxzPfBaVZ0N3ATc0GszkqTeHTEQquo7wKuTypcCW9vyVuCyrvo9VfV2VT0H7AUuTLIMOLGqHq6qAu6cNGZiX/cBayaOHiRJR8/iHscNVdU+gKral+SMVl8OPNK13VirvdOWJ9cnxrzY9jWe5HXgNOCVyT80yQY6RxkMDQ0xOjra2+SXwHXnj/c0dtDYy+A5cOBAz382B429DKb56qXXQDiUqd7Z12Hqhxvz3mLVZmAzwPDwcK1evbqHKcItd2/nxt1z3Xp/XHf+uL0MmC0jS+n1z+agGR0dtZcBNF+99HqX0cvtNBDteX+rjwFndm23Anip1VdMUT9oTJLFwEm89xSVJGme9RoIO4B1bXkdsL2rvrbdOXQWnYvHj7XTS28kuahdH7hy0piJfV0OPNSuM0iSjqIjHp8n+SNgNXB6kjHgy8BXgW1J1gMvAFcAVNWeJNuAp4Bx4Jqqerft6mo6dywtAR5oD4A7gLuS7KVzZLB2TjqTJM3IEQOhqj57iFVrDrH9JmDTFPVdwHlT1N+iBYokqX/8pLIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNcf+v2guDZjdP3qdqzZ+q9/TmBNbRpb2ewo6ijxCkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEjBA33aaZAT4j8Ai4Paq+mqfpyQd9/zm1uPLQBwhJFkE/D7wCeAc4LNJzunvrCTp+DIoRwgXAnur6ocASe4BLgWe6uusJC0YHu0cWapqXnY8o0kklwMjVfX59vpzwD+qqi9M2m4DsKG9/EXgmR5/5OnAKz2OHTT2MngWSh9gL4NqNr38QlV9eKoVg3KEkClq70mqqtoMbJ71D0t2VdXwbPczCOxl8CyUPsBeBtV89TIQ1xCAMeDMrtcrgJf6NBdJOi4NSiB8D1iV5KwkfwdYC+zo85wk6bgyEKeMqmo8yReAP6dz2+nXqmrPPP7IWZ92GiD2MngWSh9gL4NqXnoZiIvKkqT+G5RTRpKkPjMQJEnAcRgISUaSPJNkb5KN/Z5Pr5J8Lcn+JE/2ey6zkeTMJH+R5Okke5Jc2+859SrJB5I8luSvWy9f6fecZiPJoiR/leSb/Z7LbCV5PsnuJE8k2dXv+fQqyclJ7kvyg/b/zD+e0/0fT9cQ2ldk/E/gn9C51fV7wGer6pj7RHSSjwEHgDur6rx+z6dXSZYBy6rq+0k+BDwOXHaM/k4CLK2qA0neD3wXuLaqHunz1HqS5N8Aw8CJVfXpfs9nNpI8DwxX1TH9wbQkW4G/rKrb2x2Zf7eqfjJX+z/ejhB+/hUZVfUzYOIrMo45VfUd4NV+z2O2qmpfVX2/Lb8BPA0s7++selMdB9rL97fHMfmOK8kK4FPA7f2eizqSnAh8DLgDoKp+NpdhAMdfICwHXux6PcYx+pfPQpRkJfBR4NE+T6Vn7TTLE8B+4MGqOlZ7+T3gt4H/2+d5zJUCvp3k8fYVOMeivwf8H+AP2qm825PM6ZcaHW+BMK2vyNDRl+SDwDeA36yqn/Z7Pr2qqner6iN0Pm1/YZJj7nRekk8D+6vq8X7PZQ5dXFW/TOcbla9pp1yPNYuBXwZuq6qPAm8Cc3od9HgLBL8iYwC18+3fAO6uqj/u93zmQjuUHwVG+juTnlwMfKadd78H+LUkX+/vlGanql5qz/uB++mcPj7WjAFjXUed99EJiDlzvAWCX5ExYNqF2DuAp6vqd/s9n9lI8uEkJ7flJcDHgR/0dVI9qKrrq2pFVa2k8//IQ1X1G32eVs+SLG03LNBOsfw6cMzdnVdVfwu8mOQXW2kNc/xPBAzEV1ccLX34iox5k+SPgNXA6UnGgC9X1R39nVVPLgY+B+xu594BvlRV/7V/U+rZMmBru5vtfcC2qjrmb9lcAIaA+zvvPVgM/GFV/Vl/p9SzLwJ3tze0PwT++Vzu/Li67VSSdGjH2ykjSdIhGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVLz/wCoI8erFPE5aQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "  \n",
    "df = pd.read_table('/home/jovyan/rice3k/missing_by_snp.txt',delimiter='\\t')\n",
    "\n",
    "df.hist(column='N_MISS',bins=6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6730a383-bc5f-455d-9d5d-8d655cd60aba",
   "metadata": {},
   "source": [
    "#### Generate PCA using genotyping information contained in VCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7435e6e0-e31f-4a1f-acce-7de4d185e8c0",
   "metadata": {},
   "source": [
    "`Plink` alllows to create a PCA (principal components analysis) of samples, so that we can easily evaluate genetic distance between samples. \n",
    "\n",
    "This will generate a matrix of coordinates in the different component. By default, it provides the first 20 principal components of the variance-standardized relationship matrix. We will focus only the first 3 axes for subsequent visualization (`--pca 3`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3ceba829-fd6d-44dc-8ab9-e09e73ced765",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.26 64-bit (2 Apr 2022)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to plink.log.\n",
      "Options in effect:\n",
      "  --cluster\n",
      "  --matrix\n",
      "  --pca 3\n",
      "  --vcf out.vcf.gz\n",
      "\n",
      "Note: --matrix flag deprecated.  Migrate to \"--distance ibs flat-missing\",\n",
      "\"--r2 square\", etc.\n",
      "3933 MB RAM detected; reserving 1966 MB for main workspace.\n",
      "--vcf: plink-temporary.bed + plink-temporary.bim + plink-temporary.fam written.\n",
      "119202 variants loaded from .bim file.\n",
      "60 people (0 males, 0 females, 60 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to plink.nosex .\n",
      "Using 1 thread.\n",
      "Before main variant filters, 60 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.986788.\n",
      "119202 variants and 60 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Relationship matrix calculation complete.\n",
      "--pca: Results saved to plink.eigenval and plink.eigenvec .\n",
      "Distance matrix calculation complete.\n",
      "IBS matrix written to plink.mibs , and IDs to plink.mibs.id .\n",
      "Clustering... done.                        \n",
      "Cluster solution written to plink.cluster1 , plink.cluster2 , and04143454648505153555658606163656668707173757678808183858688909193959698100%\n",
      "plink.cluster3 .\n"
     ]
    }
   ],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "./plink -vcf out.vcf.gz --cluster --matrix --pca 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83c47325-bb2b-4e27-a60e-5f881e352ef7",
   "metadata": {},
   "source": [
    "#### Visualize PCA (using Pandas)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb23d0b-5d21-493f-b57e-9530ffda5930",
   "metadata": {},
   "source": [
    "With simple bash commands (`awk`), we first format a matrix dataset for Pandas, including names of sample in the first colum and coordinates X and Y. We can redirect the command to a new file called `pca.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dd0f6500-7bb4-481b-a7cc-505ae6efb967",
   "metadata": {},
   "outputs": [],
   "source": [
    "echo \"Samples PCA1 PCA2\" >pca.txt\n",
    "awk -F \" \" {'print $1\" \"$3\" \"$4'}  plink.eigenvec >>pca.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53cb7f98-09fb-4c86-a8ea-a30768333687",
   "metadata": {},
   "source": [
    "##### Download the metadata files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c24c2aff-9091-492d-9059-27257b58d7c7",
   "metadata": {},
   "source": [
    "In a second step, we want to add the genetic group information of each of the samples.\n",
    "To do so, we need to download metadata file (available here https://sniplay.southgreen.fr/examples/3K_RGP.accessions.txt) and merge it (with python) to the `pca.txt` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c2d73669-5590-4c06-979a-256f906baa74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-09-08 06:23:37--  https://sniplay.southgreen.fr/examples/3K_RGP.accessions.txt\n",
      "Resolving sniplay.southgreen.fr (sniplay.southgreen.fr)... 193.51.117.74\n",
      "Connecting to sniplay.southgreen.fr (sniplay.southgreen.fr)|193.51.117.74|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 57858 (57K) [text/plain]\n",
      "Saving to: ‘3K_RGP.accessions.txt’\n",
      "\n",
      "3K_RGP.accessions.t 100%[===================>]  56.50K  --.-KB/s    in 0.01s   \n",
      "\n",
      "2022-09-08 06:23:38 (4.61 MB/s) - ‘3K_RGP.accessions.txt’ saved [57858/57858]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wget https://sniplay.southgreen.fr/examples/3K_RGP.accessions.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4531bdcd-dd78-4d5e-a713-0a13b6698ce0",
   "metadata": {},
   "source": [
    "##### Use python/pandas to visualise (switch to python kernel)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ef6717-8a8c-4298-84f0-864c465254de",
   "metadata": {},
   "source": [
    "With a simple python script using the pandas library, we can merge the pca matrix with the metadata file (using `inner join` on the common field called \"Samples\" found in the two files to be joinned). \n",
    "\n",
    "Using the `plot.scatter` function, we can visualize the pca and colorize samples according to their genetic group (indica=ind1A in red, japx in green).\n",
    "\n",
    "Do not forget to switch to python kernell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12f1897e-717a-470e-afe3-c6d9065634c0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='PCA1', ylabel='PCA2'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZEAAAEGCAYAAACkQqisAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAnG0lEQVR4nO3deXwV5dn/8c+VPUEJILsREAQFAZUnyKYIKlURHixu8FSllp88WK1aa4Vq7eauLdZarUXr2ioWl0oVpYAbPm4EFWVRdtkXQdaQ5ST37485gZzkJCSTnDMn4ft+vc4rZ2bumVzDtF65597MOYeIiIgfSUEHICIiDZeSiIiI+KYkIiIivimJiIiIb0oiIiLiW0rQAcRCy5YtXadOnYIOQ0SkwViwYMG3zrlWtT2vUSaRTp06kZeXF3QYIiINhpl94+c8vc4SERHflERERMQ3JREREfFNSURERHxTEhEREd8aZe8sEZEG65NP4PXXoX17GDcOMjKCjqhaSiIiIoni3nth8uSD27/7HaxcmdCJRK+zREQSwbZtkQkEYONGuPvuYOKpISUREZFEUFWyWLQovnHUkpKIiEgi2LAh+v7Bg+MbRy0piYiIJILhwyE9PXJf27Zw3XXBxFNDSiIiIongiivg5pshLQ2SkuDMM2H1ajALOrJqKYmIiCQCM683VkEBFBfD3LkJ3SurjLr4iogkErOEr32Up5qIiIj4piQiIiK+KYmIiIhvSiIiIuKbkoiIiPimJCIiIr4piYiIiG9KIiIi4puSiIiI+KYkIiIivimJiIiIb0oiIiLiW6BJxMzONbOvzWyFmU2OcvwHZvZF+POBmZ0URJwiIhJdYEnEzJKBh4HzgB7AWDPrUaHYauAM51xv4HZganyjFBGR6gRZEzkVWOGcW+WcKwKmAaPKF3DOfeCc+y68+RGQE+cYRUSkGkEmkaOBdeW214f3VWU88EZVB81sgpnlmVnetm3b6ilEERGpTpBJJNqqKy5qQbOheElkUlUXc85Ndc7lOudyW7VqVU8hiohIdYJc2XA9cEy57RxgY8VCZtYbeBw4zzm3PU6xiYhIDQRZE5kPdDWzY80sDRgDzChfwMw6AC8DlzvnlgUQo4iIVCOwmohzLmRm1wKzgGTgCefcYjObGD7+KPAr4CjgEfPWHA4553KDillERCKZc1GbIRq03Nxcl5eXF3QYIiINhpkt8PNHukasi4iIb0oiIiLim5KIiIj4piQiIiK+KYmIiIhvSiIiIuKbkoiIiPimJCIiIr4piYiIiG9KIiIi4puSiIiI+KYkIiIivimJiIiIb0EuSiUiIjVRWAjPPQfbtsGQIXDqqUFHdICSiIhIIisshP79YdkyKCqCtDSYOhV+8IOgIwP0OktEJLG9+CIsXw75+RAKeT+vuaZyudLS+MeGkoiISGLbsQNKSiL37d0Lv/419OgBXbpASgokJ8ORR8L//V9cw9PrLBGRRDZkCCSV+3s/LQ3atIG774bi4siye/d65b/9FrKz4xKeaiIiIomsVy+YNs1LHBkZXpLYsaNyAikTCsFHH8UtPNVEREQS3ciRsHnzwe3MzOrLv/02nHNObGMKU01ERCTRlNUm5s2D/fsrHx82rPrzn38+NnFFoZqIiEgi2bcPBg/2uvSaQYsWXkJp2/ZgmWeega5dYft2cK7yNSo2xMeQaiIiIonkjjtg8WKvkXzPHtiwAX7yk8gyzZrBihXwpz95yaS8rCz48Y/jFm6gScTMzjWzr81shZlNjnL8BDP70MwKzeymIGIUEYmrxYu9AYZlQiFYurRyuexsuPZar8YyezYMGgSnnOIloV/8Im7hBvY6y8ySgYeBYcB6YL6ZzXDOLSlXbAdwHXBB/CMUEQlA//4wZ87BtpC0NOjbt/pzzj7b+wQgyJrIqcAK59wq51wRMA0YVb6Ac26rc24+UEVfNhGRRubnP4czz/S682ZlQe/e8Mc/Bh1VlYJsWD8aWFduez3Qz+/FzGwCMAGgQ4cOdYtMRCQoqanw7397bSHFxdCxY+RgwwQTZGQWZV+UbgY145yb6pzLdc7ltmrVqg5hiYgEzAxycuDYYxM6gUCwSWQ9cEy57RxgY0CxALBxz0aG/2M4Hf/YkRHPjWDTnk1BhiMiEl1hoTcx4xNPwKpVgYYS5Ous+UBXMzsW2ACMAf4nqGAKQ4UMemIQ63etJ+RCbNyzkdOePI2l1ywlLTktqLBERCIVFMCAAV4X37IxIjNnemNLAhBYTcQ5FwKuBWYBS4F/OucWm9lEM5sIYGZtzWw9cCPwSzNbb2ZNYxHP4m2L2Z6/nZALARAqDbF171aWbovStU5EJChPPglff+2NI9m3z/v86EeBhRPoiHXn3ExgZoV9j5b7vhnvNVfMZaRkUFRSFLFvX/G+SvtERAK1aVPlqVC2bQsmFjRi/YDCUCGuQru+wzHmxTGs3bW22nPfWP4GP//Pz3ngwwfYV7QvlmGKSGNXNlK9KoMHe11/y6Slwemnxz6uKiiJAK8te43Tnjwtaq1j1c5V9Hi4B5v3bo5yJkz5cAoXTb+I33/4e2556xb6PtaX/cVRJkwTEalOcTFccgk0b+7NlzVyZOTI9TJnn+2tJZKe7vXc6t8fnn02/vGGKYkAV79+NfnF+VUezy/OZ9qiaZX2O+e4de6tB84tCBWwbtc6Xv361ZjFKiKN1N13w+uve9OchEIwdy7cdlv0stdd573SKiiAd9/1Ek9ANIsvsLtwd7XHHY5QaajS/hJXQnFp5GD6UkrZW7S3XuOTxLBg4wIenv8wzjkm5k6kX47vsbEilb3zjrd+epn9+70EURUzb2BiwFQTAc7pcg4pSdXn04c+foh310Q+0JSkFM489kzSk9MP7DOMM489MyZxSnA+Wv8Rg58azJOfP8lTC59i6NNDmffNvKDDksakS5fIpJCSAp07BxdPDSmJAE+MeoLebXpXW2bt7rUMf244K3esjNj/4iUvMrLbSFpmtaR7y+7MumwWnZsn/oOX2rlr3l0Rrzz3h/Zz+3u3BxiRNDp33w3t28ORR3qf1q1hypSgozokvc4Cjkg7glmXzeKEP5/AzoKdlLgSMlMyKQgVVOqx9faat+nSosuB7abpTZl+yfR4hyxxtj9UubNEQagggEik0WrZ0psG/u23vUGEQ4Z4ySTBqSYS1jKrJXkT8hjdfTQDcgYwaeAkki05okwSSTRNr/1Yxy+3fEmPh3uQeWcmvf/Sm2Xbl9VX2BInV+deTVbqwW6VWalZ/Lhv/Bb+kcNEkyYwYoTXM6sBJBBQEonQqVkn/nnxP/lg/AeEXIgki/znycnOYdTxo6o4O7o9hXsY+vRQln67lIJQAYu2LuKMp87QX7ENzOjuo3l85OP0bt2bXq178cj5jzCm55igwxIJnF5nVeHP8/9MUenBcSPJlszlvS8nPSW9mrMqW7R1UUQPLodjX9E+lm9fTq82veotXom9sb3GMrbX2KDDEEkoqolUwSrMVJ9kSaQm1b47XfPM5hSXRHYDLi4ppnlm9f26S0pLuO6N6zj1sVMZ969xqrmISEJSEqnC9f2uP/AO3DAyUzO5tOeltb7O8Ucdz0U9LqJJahOSLZkmqU0Y32c8OU2rnhLMOceJj5zIQ588xPyN83lm4TPkTMmplIxERIKm11lV+NUZv6L1Ea2Zvng6LbNacvvQ2+mQ3YHv9n/Hyu9WktM0h7ZHtD3kdcyMpy94mgtOuICvvv2KXq17MaLbiGrPWbhlIV9v/zpi3/b923nuy+cYd/K4Ot2XiEh9UhKpgplxde7VXJ179YF9/1n5H0a/MJrkpGSKSor4/bDfM77PeH4x5xe8veZtWmS2YMOeDWzcs5EeLXvwwsUv0KlZJ8yM0d1H1/h378jfEXX/tvzgZuoUkVratg2WLIGjj4bjjgs6mpgx53yvSJuwcnNzXV5eXr1esyBUQKv7W0VMaZKZksmAnAF8uP7DSuMIkiyJnCNzWHHdClKTa9eWkl+cT/Y92RFTrRjGqutX0alZpzrdh4jEwZw5cMEF3qjzoiL42c/g9sQenGpmC5xzubU9T20iNbR572ZKXWnEvtSkVN755p2oA9FKXSk7Cnaw6rvaL12ZlZrFBz/6gOz0bK89JiWTV8e8qgQi0hCUlsLo0d5iUbt2eXNgTZkC9fWH7fvvw4QJcP31sHx5/VyzDvQ6q4baHtG2Uo+topKiSvvKKy4pJjsj29fv63t0X3ZO3unrXBEJ0M6dladwT072/oOfW+s/9CO98QZcdJE3UaOZt8phXh5061a369aBaiI1lJGSwUuXvEST1CY0TW9KRkoG93/vfq446YqIXlxl0pPTufLkK2vU+C4ijUjz5nDEEZH7SkqgZ8+6X/uWWw7O9Ouct4DVgw/W/bp1oJpINZxzPPbpY7yz5h26NO/CzYNuZu1P17JixwqOaXoM7Y5sR0lpCT1b92Tu6rmkJaUxc8VMnHOYGR9v+JiCUAEZKRlB34qIxIuZty7Ieed5yaOoCO66C3rVw+DiisviliWSAKlhvRrXvH4NTy18ivzifNKT0zm2+bF89r+fVZkUOj/YmdU7Vx/YzkrJYso5U/jf3P+tcyziT0lpCW+tfoudBTsZ1GEQ7Y9sH3RIcrjIz4fVq6FtWzjqqPq55v33w29/67W3gLdM7owZcNZZdb6034Z11USqsL94P1M/nXqgh1RhSSHrd69n7qq5nN/t/KjnfJv/beQ1QvvZtHdTzGMF2LZvGzO+noGZMbLbSFo1aRWX3xtPq75bxV3z7mLH/h2M6TmGS068pNryxSXFnPXMWXy2+TMMw+GYffls+uf0j1PEcljLyoITT6zfa950k9dw/9hj3vK4d9xRLwmkLpREqlBxxULw2jwKS6KseRx2WofTmLtq7oE5tzJTMzmj4xkxi7HMmp1ryJ2ae6CX2M2zbyZvQh7NMpqx6rtV5DTNoXWT1jGPI5bW7VpHn7/2YU/hHkopZdbKWWzdt5VrT722ynOe/eJZFmxaELEOyOWvXM7ynwTfo0XEFzOYNMn7JAg1rFehaXpTTutw2oFVC5MsidTk1GqTwt9H/51+Of1IsiQyUjK47+z7GHrs0JjHOmnOJHYW7CS/OJ/84nx2Fuzkh//6ITlTchj69FA6/rEjU/OmxjyOWHpm4TPsK95HKV436/zifO6cd2e156zdtZb9xZHvkDfv3RyzGEUOR4HWRMzsXOBBIBl43Dl3T4XjFj4+HMgHfuic+zSWMT224DGmfDQFw7hxwI10zO7IvLXz6JjdkUdHPMpRWVW/22yR2YL3rnyPUGmIZEvGCz/2NuzeQIkrObBd4kqYt3ZexLiWG2bdwNldzm6wqy6GSkOVxulEW/e+vAE5A8hMzTxQE0lJSiG3fR27WIpIhMBqImaWDDwMnAf0AMaaWY8Kxc4DuoY/E4C/xDKmZxY+ww2zbuCrb79i6bdLuf7N6xnZbSQrr1vJW+PeottRXl/swlAhs1fOZubymewu3F3pOilJKXFLIADndzs/YsGkzJTMSgtqpSWnNejFsC458RIyUzIPbGelZnFVn6uqPeec487h1tNvJTUpldSkVHq06sG0C6fFOlSRw0qQNZFTgRXOuVUAZjYNGAUsKVdmFPCM87qQfWRmzcysnXMuJq3VUxdMjXh/nl+cz2OfPsaFPS48sG934W76P96f9bvXY2ZkpWbx8f/7mA7ZHWIRUo3cPPBm1u1ax+OfPg7AZb0v4x9f/COiXaeopIiuLboGFWKddW/VnblXzOXm2Tezs2AnY3qNYdKgQ78XvuX0W7hp4E3kF+fTLKNZ7AMVOcwEmUSOBtaV214P9KtBmaOBSknEzCbg1Vbo0MHff9DL/6Vbpvxf+AB3vHcHq75bdaCBfV/RPq6deS0zxs7w9TvrQ3JSMo+c/wgPD38Y8CaP/O9u/82lL11KalIqhSWF3Dfsvoi14Ruifjn9ePfKd2t9XlpyGmnJaTGISESCTCLR3vdUHLRSkzLeTuemAlPBGyfiJ6BfD/k1H6z7gPyQVxtpktqEW06/JaLMsu3LInpolbgSVn630s+vq3flX6GNOH4Ea29Yy/Idy+mQ3UHjI0QkJoJMIuuBY8pt5wAbfZSpN6d1OI13r3yXqQumkmRJXJ17NSe1PSmizOCOg5m9avaB117pyekMPGZgrEKqk6Oyjqq2I4CISF0FmUTmA13N7FhgAzAG+J8KZWYA14bbS/oBu2LVHlImt31utT14ru93PXkb83hxyYuYGf1y+vHAOQ/EMiQRkYRVoyRiZqnOueIK+1o6576t6pxDcc6FzOxaYBZeF98nnHOLzWxi+PijwEy87r0r8Lr4Xun399WX5KRknrvwOR45/xGKS4ppmdUyrj2xREQSSbVzZ5nZUOBZIB34DJjgnFsTPvapc65PPIKsrVgsSiUi0pjFalGq+4BznHOt8BqtZ5tZ2cRD+vM7ig27NzDj6xl8tP4jGuPkliIi5R3qdVaac24xgHPuRTNbCrxsZpOpopfU4ezt1W8z8vmRJCclU1JawohuI3j+wuf1uktE6m7JErjuOti0Cc4915tePj096KgOmUSKzaytc24zQLjN4izgNaBhDzqIgUtfvJR9xfsObL++/HVmLp9Z5ay/IiI1snEjDBgAe/Z4a4isXu0lk+eeCzqyQ77Omgy0Kb/DObceGALcE+2Ew1WpK600FXxJaQnf7PomoIhEpNF44w0IhbwEAt7iVNOne4teBazaJOKcm+OcWxjl0JFAUWxCapiSLInjWx4fsUSumdGnXUL2PRCRRFFQAEuXwvbtVZdJTfWmgS8vKcn7BKzGEZhZSzO72szeA96hQg1F4N9j/80xTY8hIyWDtOQ07jzzTi2AJCJV+/RTOPpo6NfP+3nffdHLjRoF2dleMgFvwauf/rRyYgnAobr4Hgl8H28QYDfgFeBS51xOfMLzJ8guvqWulM17N9Mso1mlebdERCK0b++1bZTJyoJ334XcKD1tt26FO++E9eu99dvHj6/XJBKr5XG3Ap8AvwTed845M/u+nwAbspLSEm6cdSNPfv4kyUnJTB40mZsH3Ry111WSJWmeKhGprLQUNm+Gpk3hiCO8Ndi3bo0sYwZffhk9ibRuDQ8+GJ9Ya+FQr7NuATLw1vH4hZkdlj2y7njvDh7/7HH2FO1hZ8FOfvfe7/j7l38POiwRaSjWroVu3aBLF2jRAn77W8jM9BJKec5B14a1ZMOhGtYfcM71A/4bb3Dhv4D2ZjbJzLrFIb6E8NLSlyqtM/LykpcDjEhEGpSLLoI1a7xG9OJiuP9+mDMHXnnFq5VkZ3tJZeJEOO20oKOtlRo1rDvnVjnn7nTO9QL6AtnAGzGNLIFUnAk32ZJp1aRVQNGISIPzxReR3XELC2HBAjjjDG/Mx7/+BZ99Bn/4Q2Ah+lVtEjGz48xsUPl9zrkvgTeBc2IZWCL5w/f+QJPUJqQmpZKenE52Rja/HPzLoMMSkYaifYV20vR06NjR+96yJQwZAscfH/ew6sOhGtb/iNcuUlE+8AAwsr4DSkR92vVh4cSFvPLVK6QkpTCm5xjaHtE26LBEpKF4/nkYNswb1xEKwdChcOmlQUdVLw7VxXeRc65nFce+DL/eSjiaxVdEEs6WLTB/vtewPmBA9d1zZ8+GSZO8aU4uuwxuuy3mAwtj1cU3o5pjlRckFxFpiJyDGTNg+XLo3Ru+9736/x1t2sCIEYcuN3++N7hw/35v+777vNrL7bfXf0z14FCpbb6ZXVVxp5mNBxbEJiQRkThyDsaNgx/8AG65BUaP9moBQZk27WACAW88yZNPBhfPIRyqJnID8IqZ/YCDSSMXSMMbyS4N3JJtS5i7ai7NMppx8YkXk5FSXeVTpBFasgReesn7jzV4XXAffBB+9jNvgF+8ZWVBcnJkb64EmPK9KtUmEefcFmBgeIXDsraR151zb8U8Mom5WStmMfqF0ZRSSrIlc+//3cv8q+aTmao3lXIY2b794JxUZVJTYceOYJLIVVfBQw957SGlpV5SueOOg8c3b/baV447Dpo0iX98FRyqi2+Gmd0AXIg3a+9flEAaj6v+fRX5oXwKQgXsK97H6u9W89TnTwUdlkh89e4d2chtBkceCZ07BxNPhw7w+edw7bXea7ZXXoGxY71jd90FnTrB4MHehI2ffBJMjOUc6nXW00AxMA84D+iO94pLGoHvCr6L2C4IFbAtf1tA0YgEpFkzmDsXLr4Y1q3zpid5+WVISwsupk6dKs+TNX++NwFjYaH3Aa+hfsuWQGfzPVQS6VHWjdfM/oY3GaM0EkM7DWXWylkUlXhLw6SnpDO009CAoxIJQJ8+sHJl0FFUb8mSyt18d+yAffu8qVMCcqjeWcVlX5xzoRjHInH27PefZUinIaQkpdA0vSkPn/8wp3c8PeiwRCSabt28NpLysrMDbxc5VE3kJDPbHf5uQGZ42wDnnGta9amS6LIzspl12Sycc1GntReRBDJgAFx/PTzwgNdbq7TUm3Mr4P/vHqp3VnIsfqmZtQBeADoBa4BLnHPfRSn3BDAC2FrVyHmpOyUQkQbirru83lubNkH37tC8edAR1Xx53Ho2GZjrnOsKzA1vR/MUcG68ghIRSXjHHgsDByZEAoHgksgovJ5fhH9eEK2Qc+49YEecYhIRCcby5fDXv8ILLxzsedVAHKpNJFbaOOc2ATjnNplZnUf0mNkEYAJAhw4d6no5EZH4eOstGBmeED0pyXtl9dFH3iJVDUDMaiJmNsfMFkX5jIrF73POTXXO5Trnclu10oJRItJAjB/vTbmSnw9798KKFfDUU0FHVWMxq4k4586u6piZbTGzduFaSDtga1VlRUQate3bI7f37/emNmkggmoTmQGMC38fB7waUBwiIsEaPDhydHxmprfSYQMRVBK5BxhmZsuBYeFtzKy9mc0sK2RmzwMfAseb2frwFPQiIo3Hs8/CoEFee0hmJvz+997Khw1EtSsbNlRa2VBEEsrHH8OyZdCjB/zXf0UvU1LiJZKAxm35XdkwqJqIiMjh4bbb4Mwz4cc/9l5d3XNP9HLJyYGPPvdDSUREJFZWr/ZeT5X1vMrPh9/8pkE1nB+KkoiISKxs2lR5VcK0NG/69kZCSUREJFa6d688825SkrcqYSOhJCIiUhfOwYIF8PrrsHFj5LHmzeG117yfKSnQsiXMmhX49O31KahpT0REGj7n4Ec/gn/+01uXPRTypmc/u9xY68GDvQGFe/Z4y+42wMbz6iiJiIj4NWcOTJ/uNZiXueQSb8XB8sygaeNcfkmvs0RE/FqzxquNlLdzJxQVBRFNIJRERET8OuWUyG0z6NQpchqTRk5JRETEr9xcuPtuL2lkZUHbtl4D+2FESUREpC6uu85rA/nqK1i3zuvWexhRw7qISF01adKouu3WhmoiIiLim5KIiIj4piQiIpIIFi6EsWNh1ChvlHsDoTYREZGgLVrkLUy1b5+3PWcO/O1vMGZMsHHVgGoiIiJBe+SRyFHv+flw9dWwfHlwMdWQkoiISNBCoegj33Nz4ZtvAgmpppRERETiYdcur2ZRWFj52Pjx3mDFivbtg2eeiX1sdaAkIiISa48+Cm3aQJ8+0L495OVFHu/Xzxvpnpoaud85KC6OX5w+KImIiMTS4sVw441eDWTvXm90+/DhlV9fDRkCkydH1kgyMxO+cV1JREQklhYt8hakKm/XLi+ZfP45jBwJnTtD//7edPG/+pU3sePQofDWW9CjRyBh15S6+IqIxFLnzlBSErkvNdVbf33QoIO9slavhk8/hXPO8VZKbCCLVwVSEzGzFmY228yWh382j1LmGDN728yWmtliM7s+iFhFROqkb1+45hrv1VR2tjfH1vTp8Pe/R3brBa/9Y+5cb52SBiKo11mTgbnOua7A3PB2RSHgZ8657kB/4BozS+x6nYhINPfd59Uupk/3emidd17lNpEySUnRe3AlqKCSyCjg6fD3p4ELKhZwzm1yzn0a/r4HWAocHa8ARUTqVffuMGwYtGvnbY8bV7lbb3IydOgAxx0X//h8CiqJtHHObQIvWQCtqytsZp2AU4CPqykzwczyzCxv27Zt9RmriEj969ED5s2Ds86CFi2gVStv3qz33qvcEJ/AYhapmc0B2kY5dGstr3ME8BJwg3Nud1XlnHNTgakAubm5VdQTRUTi5J13YNky6NkTBg6MXqZPH2+erAYsZknEOXd2VcfMbIuZtXPObTKzdsDWKsql4iWQfzjnXo5RqCIi9euGG+Dxxw+2e0ya5HXdbYSCep01AxgX/j4OeLViATMz4G/AUufclDjGJiLi3/LlMHWqN2VJfr73uesu2LIl6MhiIqgkcg8wzMyWA8PC25hZezObGS4zCLgcONPMPg9/hgcTrohIDW3ZAmlpkfvS0qCRttUG0nrjnNsOnBVl/0ZgePj7+0DDGG0jIlLmxBOhtDRyX2oqdOkSTDwxpmlPRETqU/Pm8Oab0Lq1N+o8J8drPM/MDDqymGg4/chERBqKgQO911pFRZVfbTUyqomIiMRKI08goCQiIiJ1oCQiIiK+KYmIiIhvSiIiIuKbkoiIiPimJCIiIr4piYiIiG9KIiIi4puSiIiI+KYkIiIivimJiIiIb0oiIiLim5KIiIj4piQiIiK+KYmIiIhvSiIiIuKbkoiIiPimJCIiIr4piYiIiG9KIiIi4lsgScTMWpjZbDNbHv7ZPEqZDDP7xMwWmtliM/ttELEmkoJQAc65oMMQETkgqJrIZGCuc64rMDe8XVEhcKZz7iTgZOBcM+sfvxATx1fffkXnBzvT5K4mNLu3GW+ueDPokEREgOCSyCjg6fD3p4ELKhZwnr3hzdTw57D7M7yktITBTw5mzc41lLpSdhfu5sJ/Xsi6XeuCDk1EJLAk0sY5twkg/LN1tEJmlmxmnwNbgdnOuY+ruqCZTTCzPDPL27ZtWyxijrtSV8r5z53PtvxtuHL5MzUplU83fRpgZCIinpRYXdjM5gBtoxy6tabXcM6VACebWTPgFTPr6ZxbVEXZqcBUgNzc3EZRY3nuy+eY9828SvtDpSHaHdkugIhERCLFLIk4586u6piZbTGzds65TWbWDq+mUd21dprZO8C5QNQk0hit2LGC/aH9lfZf1OMi+rbvG0BEIiKRgnqdNQMYF/4+Dni1YgEzaxWugWBmmcDZwFfxCjARnNz2ZLJSsw5sG8YJR53Ak6OexMwCjExExBNUErkHGGZmy4Fh4W3MrL2ZzQyXaQe8bWZfAPPx2kReCyTagIw6fhRX9bmKtOQ0mqQ2oUN2B9687E0lEBFJGNYYxx3k5ua6vLy8oMOoN1v3bWV34W46NetESlLM3kCKyGHMzBY453Jre57+i9QAtG7SmtZNonZgExEJlKY9ERER35RERETENyURERHxTUlERER8UxIRERHflERERMQ3JREREfFNSUSklqYvns7Jj55Mr0d68cRnTwQdjkigNNhQpBZeW/YaP3z1h+QX5wPwkzd+QkpSClecdEXAkYkEQzURkVr464K/HkggAPnF+fwl7y8BRiQSLCURkVrISM6otC89OT2ASEQSg5KISC1MOm1SxPT8mSmZ/PqMXwcYkUiw1CYiUgu57XN5/8r3eeiThwiVhpiYO5GBxwwMOiyRwCiJiNTSKe1O4YlR6pUlAnqdJSIidaAkIiIivimJiIiIb0oiIiLim5KIiIj4piQiIiK+mXMu6BjqnZltA76J0eVbAt/G6NrxpntJXI3pfnQvianivXR0zrWq7UUaZRKJJTPLc87lBh1HfdC9JK7GdD+6l8RUX/ei11kiIuKbkoiIiPimJFJ7U4MOoB7pXhJXY7of3Utiqpd7UZuIiIj4ppqIiIj4piQiIiK+KYlEYWYtzGy2mS0P/2xeRbknzGyrmS2qsP83ZrbBzD4Pf4bHJ/KoMdb1Xmp0fjzU4l7ONbOvzWyFmU0utz/w51JVbOWOm5n9KXz8CzPrU9Nz462O97LGzL4MP4e8+EZeWQ3u5QQz+9DMCs3sptqcG291vJfaPxfnnD4VPsB9wOTw98nAvVWUGwz0ARZV2P8b4Kag76Oe7qVG5yfKvQDJwEqgM5AGLAR6JMJzqS62cmWGA28ABvQHPq7puQ3lXsLH1gAtg4rfx720BvoCd5b/31ADfS5R78Xvc1FNJLpRwNPh708DF0Qr5Jx7D9gRp5j8quu91Oj8OKlJLKcCK5xzq5xzRcC08HmJoCaxjQKecZ6PgGZm1q6G58ZTXe4l0RzyXpxzW51z84Hi2p4bZ3W5F1+URKJr45zbBBD+2drHNa4NV+GfCPIVEHW/l/r4t6gvNYnlaGBdue314X1lgnwuh4qtujI1OTee6nIvAA74j5ktMLMJMYuyZuryb9sQn0t1av1cDtvlcc1sDtA2yqFb6+HyfwFux3sgtwN/AH5UD9eNKsb3Elf1cC8WZV9ZP/a4PpcoqovtUGVqcm481eVeAAY55zaaWWtgtpl9Fa4NB6Eu/7YN8blUp9bP5bBNIs65s6s6ZmZbzKydc25TuPq9tZbX3lLuWo8Br/mPtEa/L2b3AtT1/Fqph3tZDxxTbjsH2Bi+dlyfSxRVxlaDMmk1ODee6nIvOOfKfm41s1fwXsMElURqci+xODcW6hSPn+ei11nRzQDGhb+PA16tzckV3vt+H1hUVdk4qNO91MP59akmscwHuprZsWaWBowJn5cIz6XK2MqZAVwR7tnUH9gVfnVXk3Pjyfe9mFkTMzsSwMyaAN8j2P+P1OXftiE+l6h8P5egehEk8gc4CpgLLA//bBHe3x6YWa7c88AmvAaq9cD48P5ngS+BL8IPsF0Dvpeo5yf4vQwHluH1Urm13P7An0u02ICJwMTwdwMeDh//Esg91H0F+Dx83Qtez6GF4c/iBnIvbcP/v9gN7Ax/b9pAn0vUe/H7XDTtiYiI+KbXWSIi4puSiIiI+KYkIiIivimJiIiIb0oiIiLim5KISB2YWUl4xtNFZjbdzLLC+9ua2TQzW2lmS8xsppl1K3feT82swMyyy+07yszeNrO9ZvbnIO5HpLaURETqZr9z7mTnXE+gCJhoZga8ArzjnOvinOsB3AK0KXfeWLyBYd8vt68AuA2ImJ5bJJEpiYjUn3nAccBQoNg592jZAefc5865eQBm1gU4AvglXjIpK7PPOfc+XjIRaRCURETqgZmlAOfhjczuCSyopvhYvBkC5gHHhye7E2mQlERE6ibTzD4H8oC1wN9qcM4YYJpzrhR4Gbg4duGJxNZhO4uvSD3Z75w7ufwOM1sMXBStsJn1BrriTbMN3uy8q/DmmBJpcFQTEal/bwHpZnZV2Q4z62tmZ+C9yvqNc65T+NMeONrMOgYVrEhdaAJGkTows73OuSOi7G8P/BH4L7yG8jXADcB/gPOcc1+VKzsF2OKcu9fM1uDNqJqGN8Pq95xzS2J6EyJ1oCQiIiK+6XWWiIj4piQiIiK+KYmIiIhvSiIiIuKbkoiIiPimJCIiIr4piYiIiG//Hwjz1CgJJrQIAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "  \n",
    "df1 = pd.read_table('/home/jovyan/rice3k/pca.txt',delimiter=' ')\n",
    "\n",
    "df2 = pd.read_table('/home/jovyan/rice3k/3K_RGP.accessions.txt',delimiter=';')\n",
    "\n",
    "df_cd = pd.merge(df1, df2, how='inner', left_on = 'Samples', right_on = '3K_DNA_IRIS_UNIQUE_ID')\n",
    "\n",
    "#print(df_cd)\n",
    "df_cd.to_csv('/home/jovyan/rice3k/groups', index=False)\n",
    "colors = {'ind1A':'red', 'japx':'green'}\n",
    "df_cd.plot.scatter(x='PCA1', y='PCA2',c= df_cd['TYPE'].map(colors),cmap='viridis')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e061574e-c916-43b6-985a-074b9755245d",
   "metadata": {},
   "source": [
    "#### SNP density along chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ded41a88-fa9b-40cf-9520-315695071ec6",
   "metadata": {},
   "source": [
    "We will make use of `vcftools` to calculate the density of variants along the chromosome 1 of Japonica rice, in sliding windows. To do so, we will set a 100kb sliding window to the option `--SNPdensity`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "32b1e1e9-31c3-4dfc-8357-22483627a6f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--SNPdensity 100000\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "After filtering, kept 60 out of 60 Individuals\n",
      "Outputting SNP density\n",
      "After filtering, kept 119202 out of a possible 119202 Sites\n",
      "Run Time = 0.00 seconds\n"
     ]
    }
   ],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "vcftools --gzvcf out.vcf.gz --SNPdensity 100000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e42e7ed-1e3a-46c3-8408-b2c465ccadc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\tBIN_START\tSNP_COUNT\tVARIANTS/KB\n",
      "1\t0\t382\t3.82\n",
      "1\t100000\t268\t2.68\n",
      "1\t200000\t145\t1.45\n",
      "1\t300000\t129\t1.29\n",
      "1\t400000\t277\t2.77\n",
      "1\t500000\t262\t2.62\n",
      "1\t600000\t117\t1.17\n",
      "1\t700000\t355\t3.55\n",
      "1\t800000\t9\t0.09\n"
     ]
    }
   ],
   "source": [
    "head out.snpden"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7841da73-b41c-4a7b-894c-a3422849d4a6",
   "metadata": {},
   "source": [
    "### Visualize SNP density using Circos\n",
    "\n",
    "Be carefull, this version of Circos requires to be located in the circos directory\n",
    "\n",
    "Install Circos and go in the Circos directory to run it"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38f58e3a-38ef-4e94-bee1-f1a515ba20ea",
   "metadata": {},
   "source": [
    "##### Install Circos in the terminal by typing these commands\n",
    "\n",
    "cd ~\n",
    "\n",
    "git clone https://github.com/vigsterkr/circos.git\n",
    "\n",
    "cd circos\n",
    "\n",
    "./install-unix\n",
    "\n",
    "conda create -n circos -c bioconda perl-config-general perl-gd perl-math-bezier perl-math-round perl-math-vecstat perl-params-validate perl-readonly perl-set-intspan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a03b5b15-4a82-44e9-8fbe-d4001e50f4d0",
   "metadata": {},
   "source": [
    "#### Generate input file for circos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "501ba393-a6cf-4791-928c-c76c72755eba",
   "metadata": {},
   "source": [
    "With a simple bash command (`awk`), create the input data file for Circos for visualzation of line plot (space separated format: chr start end value). The file must be named `density.tyxt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae2bb848-0072-448e-9ba0-088ad5201818",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "awk {'print $1\" \"$2\" \"$2\" \"$3'} out.snpden | grep -v BIN_START >density.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af3f863e-9f59-455d-9565-04b54453adcd",
   "metadata": {},
   "source": [
    "Go to the Circos directory and download an example of circos configuration file available at https://sniplay.southgreen.fr/examples/circos1.conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4b2c78bc-f366-4e28-8891-df78cba8481e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-09-08 07:43:33--  https://sniplay.southgreen.fr/examples/circos1.conf\n",
      "Resolving sniplay.southgreen.fr (sniplay.southgreen.fr)... 193.51.117.74\n",
      "Connecting to sniplay.southgreen.fr (sniplay.southgreen.fr)|193.51.117.74|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1893 (1.8K) [text/plain]\n",
      "Saving to: ‘circos1.conf’\n",
      "\n",
      "circos1.conf        100%[===================>]   1.85K  --.-KB/s    in 0s      \n",
      "\n",
      "2022-09-08 07:43:34 (7.62 MB/s) - ‘circos1.conf’ saved [1893/1893]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd ~/circos\n",
    "wget https://sniplay.southgreen.fr/examples/circos1.conf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ced98bbb-539d-4c92-8613-071834c4cfa0",
   "metadata": {},
   "source": [
    "##### Create a karyotype file indicating the lengths and names of chromosomes\n",
    "\n",
    "Try to guess the length of the chromosome1 to indicate in the karyotype file. For instance by using the `tail` command on `density.txt` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "100841d7-d789-427e-ba68-ec34023a9667",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 43200000 43200000 101\n"
     ]
    }
   ],
   "source": [
    "tail -1 /home/jovyan/rice3k/density.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7f91a0d-bd09-43fc-8906-30b41121e663",
   "metadata": {},
   "source": [
    "Write into a karyotype file called `karyotype.txt`, the size and color of the chromosome 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b4c47e17-a3ce-40ee-8c42-d6b6966d3c2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "echo \"chr - 1 1 0 43200000 black\" >karyotype.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6156df2-3998-474f-bbbf-80eb2438b32f",
   "metadata": {},
   "source": [
    "##### Edit the Circos configuration file to adapt the data file names. And run Circos as follows:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed24621-2784-4d87-9157-95eb39a0a9ba",
   "metadata": {},
   "source": [
    "Activate the conda environnement for Circos before running circos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aa3409f8-94e7-46af-a247-6a9d1b229f0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(circos) debug: using configuration file input circos1.conf\n",
      "debuglib: using configuration input circos1.conf\n",
      "debuglib: loading configuration from file circos1.conf\n",
      "debugload: using file from argument circos1.conf\n",
      "debugfile: looking for configuration file circos1.conf\n",
      "debugfile: found configuration file circos1.conf\n",
      "zoomregion ideogram 0 chr 1         0  43200000 scale  1.00 absolutescale  1.00\n",
      "registering tag 1\n",
      "ideogramspacing 1 1 1 1 1080.000025\n",
      "ideogramreport   0     1   0     1      0.000  43200.000  43200.001       0.000       0.000 r 1350 1340 1350 10\n",
      "drawing ticks 1 radius 1350 type absolute spacing 1000000\n",
      "drawing ticks 1 radius 1350 type absolute spacing 100000\n",
      "drawing plot type histogram at z-depth 0\n",
      "created image at ./circos.png\n",
      "(circos) "
     ]
    }
   ],
   "source": [
    "cd ~/circos\n",
    "conda activate circos\n",
    "bin/circos --conf circos1.conf\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1d19fbc-4e4e-4d18-ae61-3ccb3001eb39",
   "metadata": {},
   "source": [
    "### Comparison of populations using FST"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95f295c3-aac8-4a01-90bd-31e8f14d944d",
   "metadata": {},
   "source": [
    "FST is an index that reflect the level of differenciation between populations. We will calculate FST values for each variant in order to know if they can dissociate specific alleles of the two population ind1B and japx.\n",
    "\n",
    "Using `grep` and `awk`, create two distinct file (called `pop1` and `pop2`) listing the names of accessions that are assigned to each group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "969ee4ec-888b-45e8-baed-9edec54cb0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/jovyan/rice3k\n",
    "grep ind1A 3K_RGP.accessions.txt | awk -F \";\" {'print $1'} >pop1\n",
    "grep japx 3K_RGP.accessions.txt | awk -F \";\" {'print $1'} >pop2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bcf271c-c5cc-4f90-a4b1-b14892daf293",
   "metadata": {},
   "source": [
    "We will use `vcftools` for the calculation of FST. It can be calculated by specifying population files to the option `--weir-fst-pop`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "11440162-3c0e-4ea9-bc77-6722d2e5c0b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--weir-fst-pop pop1\n",
      "\t--weir-fst-pop pop2\n",
      "\t--keep pop1\n",
      "\t--keep pop2\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 60 out of 60 Individuals\n",
      "Outputting Weir and Cockerham Fst estimates.\n",
      "Weir and Cockerham mean Fst estimate: 0.46204\n",
      "Weir and Cockerham weighted Fst estimate: 0.6612\n",
      "After filtering, kept 119202 out of a possible 119202 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcftools --gzvcf out.vcf.gz --weir-fst-pop pop1 --weir-fst-pop pop2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3760e19d-5a53-4582-a315-a875c526f2ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t43254106\t0.649498\n",
      "1\t43254132\t0.748866\n",
      "1\t43254181\t0.560328\n",
      "1\t43254300\t0.286511\n",
      "1\t43254513\t0.445836\n",
      "1\t43255505\t0.141379\n",
      "1\t43255617\t0.141379\n",
      "1\t43269402\t0.582643\n",
      "1\t43269449\t0.255167\n",
      "1\t43269550\t0.35192\n"
     ]
    }
   ],
   "source": [
    "tail out.weir.fst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a025f830-1f6d-4d1e-a5e5-88ad008b6a64",
   "metadata": {},
   "source": [
    "Now, we want to look at the average of FST by sliding window along the chromosome, in order to detect potential regions showing a high level of differenciation between the two genetic groups. \n",
    "\n",
    "We can use the option `--fst-window-size` and define a 100kb size of window."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9511a323-e87f-400d-be04-78f8ce1543e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--fst-window-size 100000\n",
      "\t--weir-fst-pop pop1\n",
      "\t--weir-fst-pop pop2\n",
      "\t--keep pop1\n",
      "\t--keep pop2\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 60 out of 60 Individuals\n",
      "Outputting Windowed Weir and Cockerham Fst estimates.\n",
      "Weir and Cockerham mean Fst estimate: 0.46204\n",
      "Weir and Cockerham weighted Fst estimate: 0.6612\n",
      "After filtering, kept 119202 out of a possible 119202 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcftools --gzvcf out.vcf.gz --weir-fst-pop pop1 --weir-fst-pop pop2 --fst-window-size 100000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d5ce094-d237-412b-932d-1e30a85a6854",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\tBIN_START\tBIN_END\tN_VARIANTS\tWEIGHTED_FST\tMEAN_FST\n",
      "1\t1\t100000\t382\t0.622881\t0.549616\n",
      "1\t100001\t200000\t268\t0.603637\t0.491764\n",
      "1\t200001\t300000\t145\t0.780704\t0.611452\n",
      "1\t300001\t400000\t129\t0.88753\t0.774335\n",
      "1\t400001\t500000\t277\t0.815848\t0.711635\n",
      "1\t500001\t600000\t262\t0.744546\t0.59308\n",
      "1\t600001\t700000\t117\t0.743666\t0.601746\n",
      "1\t700001\t800000\t355\t0.310809\t0.282787\n",
      "1\t800001\t900000\t9\t0.47269\t0.300306\n"
     ]
    }
   ],
   "source": [
    "head out.windowed.weir.fst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2d189cc1-7696-4d71-b416-ca5cb4c20131",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(circos) (circos) (circos) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd ~\n",
    "awk {'print $1\" \"$2\" \"$2\" \"$3'} out.weir.fst | grep -v FST >fst.txt\n",
    "awk {'print $1\" \"$2\" \"$2\" \"$6'} out.windowed.weir.fst | grep -v FST >fst_window.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b400e09-5673-4119-b3c2-f08bcbe9cb4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate circos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71eb968-3535-4fbe-a93a-cb9168fd90f4",
   "metadata": {},
   "source": [
    "#### Add a new track in Circos for FST means in sliding windows\n",
    "\n",
    "Modify the Circos configuration file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "dfa6d2cc-350a-433b-9ddd-2fc21c07f2ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(circos) debug: using configuration file input circos1.conf\n",
      "debuglib: using configuration input circos1.conf\n",
      "debuglib: loading configuration from file circos1.conf\n",
      "debugload: using file from argument circos1.conf\n",
      "debugfile: looking for configuration file circos1.conf\n",
      "debugfile: found configuration file circos1.conf\n",
      "zoomregion ideogram 0 chr 1         0  46000000 scale  1.00 absolutescale  1.00\n",
      "registering tag 1\n",
      "ideogramspacing 1 1 1 1 1150.000025\n",
      "ideogramreport   0     1   0     1      0.000  46000.000  46000.001       0.000       0.000 r 1350 1340 1350 10\n",
      "drawing ticks 1 radius 1350 type absolute spacing 1000000\n",
      "drawing ticks 1 radius 1350 type absolute spacing 100000\n",
      "drawing plot type histogram at z-depth 0\n",
      "drawing plot type line at z-depth 0\n",
      "drawing plot type line at z-depth 0\n",
      "drawing plot type histogram at z-depth 0\n",
      "created image at ./circos.png\n",
      "(circos) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd ~/circos\n",
    "bin/circos -conf circos1.conf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31bfde0a-aae3-4823-b579-d8a620b8d70a",
   "metadata": {},
   "source": [
    "#### Compare SNP density between the 2 phylogroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "78c193ea-c821-489c-9c47-d410657c6c7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(circos) \n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--keep pop1\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.01\n",
      "\t--out pop1\n",
      "\t--recode\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 30 out of 60 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 82263 out of a possible 119202 Sites\n",
      "Run Time = 3.00 seconds\n",
      "(circos) \n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf out.vcf.gz\n",
      "\t--keep pop2\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.01\n",
      "\t--out pop2\n",
      "\t--recode\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 30 out of 60 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 97253 out of a possible 119202 Sites\n",
      "Run Time = 2.00 seconds\n",
      "(circos) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd ~\n",
    "vcftools --keep pop1 --gzvcf out.vcf.gz --out pop1 --recode --recode-INFO-all --maf 0.01\n",
    "vcftools --keep pop2 --gzvcf out.vcf.gz --out pop2 --recode --recode-INFO-all --maf 0.01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8ab759f1-15bc-4c23-8ad2-d821501f173e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf pop1.recode.vcf\n",
      "\t--out pop1\n",
      "\t--SNPdensity 100000\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "After filtering, kept 30 out of 30 Individuals\n",
      "Outputting SNP density\n",
      "After filtering, kept 82263 out of a possible 82263 Sites\n",
      "Run Time = 0.00 seconds\n",
      "(circos) \n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf pop2.recode.vcf\n",
      "\t--out pop2\n",
      "\t--SNPdensity 100000\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PR,Number=0,Type=Flag,Description=\"Provisional reference allele, may not be based on real reference genome\">\n",
      "After filtering, kept 30 out of 30 Individuals\n",
      "Outputting SNP density\n",
      "After filtering, kept 97253 out of a possible 97253 Sites\n",
      "Run Time = 1.00 seconds\n",
      "(circos) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "vcftools --gzvcf pop1.recode.vcf --SNPdensity 100000 --out pop1\n",
    "vcftools --gzvcf pop2.recode.vcf --SNPdensity 100000 --out pop2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "968895f3-13b3-4f77-a163-f389ccf3b62b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(circos) (circos) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "awk {'print $1\" \"$2\" \"$2\" \"$3'} pop1.snpden | grep -v BIN_START >pop1.density.txt\n",
    "awk {'print $1\" \"$2\" \"$2\" \"$3'} pop2.snpden | grep -v BIN_START >pop2.density.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d793ef56-3c04-4992-8781-d3cf8b227e2b",
   "metadata": {},
   "source": [
    "#### Add a new track in Circos for SNP density for the 2 populations in the same track\n",
    "\n",
    "Modify the Circos configuration file (add the type line for this)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33227c09-e983-4c7e-b333-54107c404f13",
   "metadata": {
    "tags": []
   },
   "source": [
    "<img src=\"circos.png\" align=\"center\" width=\"70%\" style=\"display: block; margin: auto;\"/> "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3ab966-5dba-4641-9afe-98707f0f7631",
   "metadata": {},
   "source": [
    "#### Calculate the ratio transition/transversion (using vcftools option --TsTv)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc-showcode": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
